{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "solution": "hidden"
   },
   "source": [
    "<img src=\"https://github.com/i3hsInnovation/resources/blob/master/images/introbanner.png?raw=true\" />"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<table style=\"float:right\" width=\"50%\">\n",
    "    <tr>\n",
    "        <td>                      \n",
    "            <div style=\"text-align: left\"><a href=\"\" target=\"_blank\">Dr Peter Causey-Freeman</a></div>\n",
    "            <div style=\"text-align: left\">Lecturer - Healthcare Sciences</div>\n",
    "            <div style=\"text-align: left\">(Clinical Bioinformatics)</div>\n",
    "            <div style=\"text-align: left\">The University of Manchester</div>\n",
    "         </td>\n",
    "         <td>\n",
    "             <img src=\"https://github.com/i3hsInnovation/resources/blob/master/images/pete.001.png?raw=true\" width=\"40%\" />\n",
    "         </td>\n",
    "     </tr>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to SPRINT 1 Notebook A\n",
    "\n",
    "****\n",
    "\n",
    "#### About this Notebook\n",
    "This introductory notebook, and subsequent SPRINT_introduction notebooks will begin to introduce the concept of retrieving data from Application Programming Interfaces [API](https://en.wikipedia.org/wiki/Application_programming_interface) which are web-hosted ([web API](https://en.wikipedia.org/wiki/Web_API))\n",
    "\n",
    "Many bioinformatics tools and data repositories can be accessed using web APIs including NCBI and Ensembl. \n",
    "\n",
    "Although we cannot hope to demonstrate how each an every useful bioinformatics web API works during this 10 week course, we will give you a broad overview of the tools we use to request data from these resources and the tools we use to make-sense of the data that are returned \n",
    "\n",
    "This notebook is at <code>Beginner</code> level and will take approximately 2 - 3 hours to complete."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<div class=\"alert alert-block alert-warning\"><b>Learning Objective:</b><ul><li> Connect and create Python functions and application programming interfaces (APIs) to analyse biological data</li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"top\"></a>\n",
    "\n",
    "### Table of Contents\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- [1 Accessing RefSeq records with biopython](#biopython)\n",
    "- [2 Installing biopython](#install)\n",
    "- [3 So what exactly is biopython](#overview)\n",
    "- [4 The Bio.Entrez module](#bio_entrez)\n",
    "- [5 SeqIO, SeqRecord and Seq](#seq)\n",
    "- [6 Bio.Entrez, SeqIO and Seq example](#be_example)\n",
    "- [7 Sanity checking with biopython](#sanity)\n",
    "- [8 Summary and Assignment](#summary)\n",
    "\n",
    "------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"biopython\"></a>\n",
    "<table width=\"100%\" style=\"float:left\">\n",
    "    <tr>\n",
    "        <td width=\"60%\" style=\"text-align: left\">\n",
    "            <h2>1 Accessing RefSeq records with <a href=\"https://biopython.org/\" target=\"_blank\">biopython</a></h2>\n",
    "        </td>\n",
    "        <td width=\"40%\">\n",
    "            <img src=\"data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiIHN0YW5kYWxvbmU9Im5vIj8+Cjwh%0D%0ALS0gQ3JlYXRlZCB3aXRoIElua3NjYXBlIChodHRwOi8vd3d3Lmlua3NjYXBlLm9yZy8pIC0tPgoK%0D%0APHN2ZwogICB4bWxuczpkYz0iaHR0cDovL3B1cmwub3JnL2RjL2VsZW1lbnRzLzEuMS8iCiAgIHht%0D%0AbG5zOmNjPSJodHRwOi8vY3JlYXRpdmVjb21tb25zLm9yZy9ucyMiCiAgIHhtbG5zOnJkZj0iaHR0%0D%0AcDovL3d3dy53My5vcmcvMTk5OS8wMi8yMi1yZGYtc3ludGF4LW5zIyIKICAgeG1sbnM6c3ZnPSJo%0D%0AdHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIKICAgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIw%0D%0AMDAvc3ZnIgogICB4bWxuczpzb2RpcG9kaT0iaHR0cDovL3NvZGlwb2RpLnNvdXJjZWZvcmdlLm5l%0D%0AdC9EVEQvc29kaXBvZGktMC5kdGQiCiAgIHhtbG5zOmlua3NjYXBlPSJodHRwOi8vd3d3Lmlua3Nj%0D%0AYXBlLm9yZy9uYW1lc3BhY2VzL2lua3NjYXBlIgogICB3aWR0aD0iMTgwMCIKICAgaGVpZ2h0PSIx%0D%0AMjAwIgogICB2aWV3Qm94PSIwIDAgMTgwMCAxMjAwIgogICBpZD0ic3ZnMiIKICAgdmVyc2lvbj0i%0D%0AMS4xIgogICBpbmtzY2FwZTp2ZXJzaW9uPSIwLjkxIHIxMzcyNSIKICAgc29kaXBvZGk6ZG9jbmFt%0D%0AZT0iYmlvcHl0aG9uX2xvZ28uc3ZnIgogICBpbmtzY2FwZTpleHBvcnQtZmlsZW5hbWU9IkY6XElu%0D%0AZm9ybWF0aWtcYmlvcHl0aG9uX3dlYnNpdGVcYXNzZXRzXGltYWdlc1xiaW9weXRob25fbG9nb194%0D%0Acy5wbmciCiAgIGlua3NjYXBlOmV4cG9ydC14ZHBpPSIxNSIKICAgaW5rc2NhcGU6ZXhwb3J0LXlk%0D%0AcGk9IjE1Ij4KICA8ZGVmcwogICAgIGlkPSJkZWZzNCI+CiAgICA8cGF0dGVybgogICAgICAgeT0i%0D%0AMCIKICAgICAgIHg9IjAiCiAgICAgICBoZWlnaHQ9IjYiCiAgICAgICB3aWR0aD0iNiIKICAgICAg%0D%0AIHBhdHRlcm5Vbml0cz0idXNlclNwYWNlT25Vc2UiCiAgICAgICBpZD0iRU1GaGJhc2VwYXR0ZXJu%0D%0AIiAvPgogIDwvZGVmcz4KICA8c29kaXBvZGk6bmFtZWR2aWV3CiAgICAgaWQ9ImJhc2UiCiAgICAg%0D%0AcGFnZWNvbG9yPSIjZmZmZmZmIgogICAgIGJvcmRlcmNvbG9yPSIjNjY2NjY2IgogICAgIGJvcmRl%0D%0Acm9wYWNpdHk9IjEuMCIKICAgICBpbmtzY2FwZTpwYWdlb3BhY2l0eT0iMC4wIgogICAgIGlua3Nj%0D%0AYXBlOnBhZ2VzaGFkb3c9IjIiCiAgICAgaW5rc2NhcGU6em9vbT0iMC41MDAwMDAwMSIKICAgICBp%0D%0AbmtzY2FwZTpjeD0iMTAwMS4xNTMxIgogICAgIGlua3NjYXBlOmN5PSI1NzkuNDAwMTkiCiAgICAg%0D%0AaW5rc2NhcGU6ZG9jdW1lbnQtdW5pdHM9InB4IgogICAgIGlua3NjYXBlOmN1cnJlbnQtbGF5ZXI9%0D%0AImxheWVyMSIKICAgICBzaG93Z3JpZD0iZmFsc2UiCiAgICAgaW5rc2NhcGU6d2luZG93LXdpZHRo%0D%0APSIxNjgwIgogICAgIGlua3NjYXBlOndpbmRvdy1oZWlnaHQ9Ijk4NyIKICAgICBpbmtzY2FwZTp3%0D%0AaW5kb3cteD0iLTgiCiAgICAgaW5rc2NhcGU6d2luZG93LXk9Ii04IgogICAgIGlua3NjYXBlOndp%0D%0AbmRvdy1tYXhpbWl6ZWQ9IjEiCiAgICAgdW5pdHM9InB4IgogICAgIHNob3dndWlkZXM9ImZhbHNl%0D%0AIgogICAgIGlua3NjYXBlOmd1aWRlLWJib3g9InRydWUiPgogICAgPHNvZGlwb2RpOmd1aWRlCiAg%0D%0AICAgICBwb3NpdGlvbj0iODk5Ljk5OTk4LDYzNi45OTk5OCIKICAgICAgIG9yaWVudGF0aW9uPSIx%0D%0ALDAiCiAgICAgICBpZD0iZ3VpZGU0MTUwIiAvPgogIDwvc29kaXBvZGk6bmFtZWR2aWV3PgogIDxt%0D%0AZXRhZGF0YQogICAgIGlkPSJtZXRhZGF0YTciPgogICAgPHJkZjpSREY+CiAgICAgIDxjYzpXb3Jr%0D%0ACiAgICAgICAgIHJkZjphYm91dD0iIj4KICAgICAgICA8ZGM6Zm9ybWF0PmltYWdlL3N2Zyt4bWw8%0D%0AL2RjOmZvcm1hdD4KICAgICAgICA8ZGM6dHlwZQogICAgICAgICAgIHJkZjpyZXNvdXJjZT0iaHR0%0D%0AcDovL3B1cmwub3JnL2RjL2RjbWl0eXBlL1N0aWxsSW1hZ2UiIC8+CiAgICAgICAgPGRjOnRpdGxl%0D%0AIC8+CiAgICAgIDwvY2M6V29yaz4KICAgIDwvcmRmOlJERj4KICA8L21ldGFkYXRhPgogIDxnCiAg%0D%0AICAgaW5rc2NhcGU6bGFiZWw9IkxheWVyIDEiCiAgICAgaW5rc2NhcGU6Z3JvdXBtb2RlPSJsYXll%0D%0AciIKICAgICBpZD0ibGF5ZXIxIgogICAgIHRyYW5zZm9ybT0idHJhbnNsYXRlKDAsMTQ3LjYzNzgy%0D%0AKSI+CiAgICA8ZwogICAgICAgc3R5bGU9ImZvbnQtc3R5bGU6bm9ybWFsO2ZvbnQtdmFyaWFudDpu%0D%0Ab3JtYWw7Zm9udC13ZWlnaHQ6bm9ybWFsO2ZvbnQtc3RyZXRjaDpub3JtYWw7Zm9udC1zaXplOjMx%0D%0AMi41cHg7bGluZS1oZWlnaHQ6MTI1JTtmb250LWZhbWlseTpDb21mb3J0YWE7LWlua3NjYXBlLWZv%0D%0AbnQtc3BlY2lmaWNhdGlvbjonQ29tZm9ydGFhLCBOb3JtYWwnO3RleHQtYWxpZ246c3RhcnQ7bGV0%0D%0AdGVyLXNwYWNpbmc6MHB4O3dvcmQtc3BhY2luZzowcHg7d3JpdGluZy1tb2RlOmxyLXRiO3RleHQt%0D%0AYW5jaG9yOnN0YXJ0O2ZpbGw6IzY3Njc2NztmaWxsLW9wYWNpdHk6MTtzdHJva2U6bm9uZTtzdHJv%0D%0Aa2Utd2lkdGg6MXB4O3N0cm9rZS1saW5lY2FwOmJ1dHQ7c3Ryb2tlLWxpbmVqb2luOm1pdGVyO3N0%0D%0Acm9rZS1vcGFjaXR5OjEiCiAgICAgICBpZD0idGV4dDQzMjYiPgogICAgICA8cGF0aAogICAgICAg%0D%0AICBkPSJtIDE3OS4yOTI1NCw3NDkuNTc3MjEgcSAyMy43NSwwIDQyLjgxMjUsMTEuNTYyNSAxOS4z%0D%0ANzUsMTEuMjUgMzAuMzEyNSwzMS4yNSAxMS4yNSwxOS42ODc1IDExLjI1LDQ0LjA2MjUgMCwyNC4z%0D%0ANzUgLTExLjU2MjUsNDQuMzc1IC0xMS4yNSwyMCAtMzEuMjUsMzEuNTYyNSAtMTkuNjg3NSwxMS4y%0D%0ANSAtNDQuMDYyNSwxMS4yNSAtMjQuMzc1LDAgLTQ0LjM3NSwtMTEuMjUgLTE5LjY4NzUsLTExLjU2%0D%0AMjUgLTMxLjI1LC0zMS41NjI1IC0xMS4yNTAwMDIsLTIwIC0xMS4yNTAwMDIsLTQ0LjM3NSBsIDAs%0D%0ALTczLjQzNzUgcSAwLC0xLjg3NSAwLjMxMjUsLTIuODEyNSBsIDAsLTcwIHEgMCwtNS4zMTI1IDMu%0D%0AMTI1LC04Ljc1IDMuNDM3NSwtMy40Mzc1IDguNzUwMDAyLC0zLjQzNzUgNS4zMTI1LDAgOC43NSwz%0D%0ALjQzNzUgMy40Mzc1LDMuNDM3NSAzLjQzNzUsOC43NSBsIDAsOTEuNTYyNSBxIDEwLjkzNzUsLTE1%0D%0AIDI3LjgxMjUsLTIzLjQzNzUgMTYuODc1LC04Ljc1IDM3LjE4NzUsLTguNzUgeiBtIC0yLjUsMTUy%0D%0ALjE4NzUgcSAxNy44MTI1LDAgMzIuMTg3NSwtOC40Mzc1IDE0LjY4NzUsLTguNzUgMjIuODEyNSwt%0D%0AMjMuNDM3NSA4LjQzNzUsLTE1IDguNDM3NSwtMzMuNDM3NSAwLC0xOC40Mzc1IC04LjQzNzUsLTMz%0D%0ALjEyNSAtOC4xMjUsLTE1IC0yMi44MTI1LC0yMy40Mzc1IC0xNC4zNzUsLTguNDM3NSAtMzIuMTg3%0D%0ANSwtOC40Mzc1IC0yNC42ODc1LDAgLTQyLjE4NzUsMTUuNjI1IC0xNy4xODc1LDE1LjYyNSAtMjAu%0D%0AMzEyNSw0MCAwLDAuOTM3NSAtMC4zMTI1LDEuNTYyNSAtMC4zMTI1LDIuNSAtMC4zMTI1LDcuODEy%0D%0ANSAwLDE4LjQzNzUgOC4xMjUsMzMuNDM3NSA4LjQzNzUsMTQuNjg3NSAyMi44MTI1LDIzLjQzNzUg%0D%0AMTQuMzc1LDguNDM3NSAzMi4xODc1LDguNDM3NSB6IgogICAgICAgICBzdHlsZT0iZm9udC1zdHls%0D%0AZTpub3JtYWw7Zm9udC12YXJpYW50Om5vcm1hbDtmb250LXdlaWdodDpub3JtYWw7Zm9udC1zdHJl%0D%0AdGNoOm5vcm1hbDtmb250LXNpemU6MzEyLjVweDtsaW5lLWhlaWdodDoxMjUlO2ZvbnQtZmFtaWx5%0D%0AOkNvbWZvcnRhYTstaW5rc2NhcGUtZm9udC1zcGVjaWZpY2F0aW9uOidDb21mb3J0YWEsIE5vcm1h%0D%0AbCc7dGV4dC1hbGlnbjpzdGFydDt3cml0aW5nLW1vZGU6bHItdGI7dGV4dC1hbmNob3I6c3RhcnQi%0D%0ACiAgICAgICAgIGlkPSJwYXRoNDE0OCIgLz4KICAgICAgPHBhdGgKICAgICAgICAgZD0ibSAzMjgu%0D%0AMDUyMyw5MjIuMDc3MjEgcSAtNS4zMTI1LDAgLTguNzUsLTMuNDM3NSAtMy40Mzc1LC0zLjQzNzUg%0D%0ALTMuNDM3NSwtOC43NSBsIDAsLTE0Ni41NjI1IHEgMCwtNS4zMTI1IDMuNDM3NSwtOC43NSAzLjQz%0D%0ANzUsLTMuNDM3NSA4Ljc1LC0zLjQzNzUgNS4zMTI1LDAgOC40Mzc1LDMuNDM3NSAzLjQzNzUsMy40%0D%0AMzc1IDMuNDM3NSw4Ljc1IGwgMCwxNDYuNTYyNSBxIDAsNS4zMTI1IC0zLjQzNzUsOC43NSAtMy4x%0D%0AMjUsMy40Mzc1IC04LjQzNzUsMy40Mzc1IHogbSAtMC4zMTI1LC0yMDMuNDM3NSBxIC03LjE4NzUs%0D%0AMCAtMTIuMTg3NSwtNSAtNC42ODc1LC01IC00LjY4NzUsLTEyLjE4NzUgMCwtNy41IDUsLTEyLjE4%0D%0ANzUgNSwtNC42ODc1IDEyLjE4NzUsLTQuNjg3NSA2Ljg3NSwwIDExLjg3NSw0LjY4NzUgNSw0LjY4%0D%0ANzUgNSwxMi4xODc1IDAsNy4xODc1IC01LDEyLjE4NzUgLTUsNSAtMTIuMTg3NSw1IHoiCiAgICAg%0D%0AICAgIHN0eWxlPSJmb250LXN0eWxlOm5vcm1hbDtmb250LXZhcmlhbnQ6bm9ybWFsO2ZvbnQtd2Vp%0D%0AZ2h0Om5vcm1hbDtmb250LXN0cmV0Y2g6bm9ybWFsO2ZvbnQtc2l6ZTozMTIuNXB4O2xpbmUtaGVp%0D%0AZ2h0OjEyNSU7Zm9udC1mYW1pbHk6Q29tZm9ydGFhOy1pbmtzY2FwZS1mb250LXNwZWNpZmljYXRp%0D%0Ab246J0NvbWZvcnRhYSwgTm9ybWFsJzt0ZXh0LWFsaWduOnN0YXJ0O3dyaXRpbmctbW9kZTpsci10%0D%0AYjt0ZXh0LWFuY2hvcjpzdGFydCIKICAgICAgICAgaWQ9InBhdGg0MTUwIiAvPgogICAgICA8cGF0%0D%0AaAogICAgICAgICBkPSJtIDQ3OS4wMDkzMyw5MjMuNjM5NzEgcSAtMjQuNjg3NSwwIC00NC42ODc1%0D%0ALC0xMS4yNSAtMTkuNjg3NSwtMTEuMjUgLTMwLjkzNzUsLTMwLjkzNzUgLTExLjI1LC0yMCAtMTEu%0D%0AMjUsLTQ0LjY4NzUgMCwtMjQuNjg3NSAxMS4yNSwtNDQuNjg3NSAxMS4yNSwtMjAgMzAuOTM3NSwt%0D%0AMzEuMjUgMjAsLTExLjI1IDQ0LjY4NzUsLTExLjI1IDI0LjY4NzUsMCA0NC4zNzUsMTEuMjUgMTku%0D%0ANjg3NSwxMS4yNSAzMC45Mzc1LDMxLjI1IDExLjU2MjUsMjAgMTEuNTYyNSw0NC42ODc1IDAsMjQu%0D%0ANjg3NSAtMTEuMjUsNDQuNjg3NSAtMTEuMjUsMTkuNjg3NSAtMzEuMjUsMzAuOTM3NSAtMTkuNjg3%0D%0ANSwxMS4yNSAtNDQuMzc1LDExLjI1IHogbSAwLC0yMS44NzUgcSAxOC4xMjUsMCAzMi41LC04LjQz%0D%0ANzUgMTQuMzc1LC04LjQzNzUgMjIuNSwtMjMuMTI1IDguMTI1LC0xNSA4LjEyNSwtMzMuNDM3NSAw%0D%0ALC0xOC43NSAtOC4xMjUsLTMzLjQzNzUgLTguMTI1LC0xNSAtMjIuNSwtMjMuNDM3NSAtMTQuMzc1%0D%0ALC04LjQzNzUgLTMyLjUsLTguNDM3NSAtMTcuODEyNSwwIC0zMi41LDguNDM3NSAtMTQuMzc1LDgu%0D%0ANDM3NSAtMjIuODEyNSwyMy40Mzc1IC04LjEyNSwxNC42ODc1IC04LjEyNSwzMy40Mzc1IDAsMTgu%0D%0ANDM3NSA4LjEyNSwzMy40Mzc1IDguNDM3NSwxNC42ODc1IDIyLjgxMjUsMjMuMTI1IDE0LjY4NzUs%0D%0AOC40Mzc1IDMyLjUsOC40Mzc1IHoiCiAgICAgICAgIHN0eWxlPSJmb250LXN0eWxlOm5vcm1hbDtm%0D%0Ab250LXZhcmlhbnQ6bm9ybWFsO2ZvbnQtd2VpZ2h0Om5vcm1hbDtmb250LXN0cmV0Y2g6bm9ybWFs%0D%0AO2ZvbnQtc2l6ZTozMTIuNXB4O2xpbmUtaGVpZ2h0OjEyNSU7Zm9udC1mYW1pbHk6Q29tZm9ydGFh%0D%0AOy1pbmtzY2FwZS1mb250LXNwZWNpZmljYXRpb246J0NvbWZvcnRhYSwgTm9ybWFsJzt0ZXh0LWFs%0D%0AaWduOnN0YXJ0O3dyaXRpbmctbW9kZTpsci10Yjt0ZXh0LWFuY2hvcjpzdGFydCIKICAgICAgICAg%0D%0AaWQ9InBhdGg0MTUyIiAvPgogICAgICA8cGF0aAogICAgICAgICBkPSJtIDY5OS44NjM4Myw3NDku%0D%0ANTc3MjEgcSAyNC4zNzUsMCA0NC4wNjI1LDExLjU2MjUgMjAsMTEuMjUgMzEuMjUsMzEuMjUgMTEu%0D%0ANTYyNSwyMCAxMS41NjI1LDQ0LjM3NSAwLDI0LjM3NSAtMTEuMjUsNDQuMzc1IC0xMC45Mzc1LDE5%0D%0ALjY4NzUgLTMwLjMxMjUsMzEuMjUgLTE5LjA2MjUsMTEuMjUgLTQyLjgxMjUsMTEuMjUgLTIwLjMx%0D%0AMjUsMCAtMzcuMTg3NSwtOC40Mzc1IC0xNi44NzUsLTguNzUgLTI3LjgxMjUsLTIzLjc1IGwgMCw5%0D%0AMS41NjI1IHEgMCw1LjMxMjUgLTMuNDM3NSw4Ljc1IC0zLjQzNzUsMy40Mzc1IC04Ljc1LDMuNDM3%0D%0ANSAtNS4zMTI1LDAgLTguNzUsLTMuNDM3NSAtMy4xMjUsLTMuNDM3NSAtMy4xMjUsLTguNzUgbCAw%0D%0ALC03MCBxIC0wLjMxMjUsLTAuOTM3NSAtMC4zMTI1LC0yLjgxMjUgbCAwLC03My40Mzc1IHEgMCwt%0D%0AMjQuMzc1IDExLjI1LC00NC4zNzUgMTEuNTYyNSwtMjAgMzEuMjUsLTMxLjI1IDIwLC0xMS41NjI1%0D%0AIDQ0LjM3NSwtMTEuNTYyNSB6IG0gMCwxNTIuMTg3NSBxIDE3LjgxMjUsMCAzMi4xODc1LC04LjQz%0D%0ANzUgMTQuNjg3NSwtOC40Mzc1IDIyLjgxMjUsLTIzLjEyNSA4LjQzNzUsLTE1IDguNDM3NSwtMzMu%0D%0ANDM3NSAwLC0xOC40Mzc1IC04LjQzNzUsLTMzLjEyNSAtOC4xMjUsLTE1IC0yMi44MTI1LC0yMy40%0D%0AMzc1IC0xNC4zNzUsLTguNzUgLTMyLjE4NzUsLTguNzUgLTE3LjgxMjUsMCAtMzIuMTg3NSw4Ljc1%0D%0AIC0xNC4zNzUsOC40Mzc1IC0yMi44MTI1LDIzLjQzNzUgLTguMTI1LDE0LjY4NzUgLTguMTI1LDMz%0D%0ALjEyNSAwLDUuMzEyNSAwLjMxMjUsNy44MTI1IDAuMzEyNSwwLjYyNSAwLjMxMjUsMS41NjI1IDMu%0D%0AMTI1LDI0LjM3NSAyMC4zMTI1LDQwIDE3LjUsMTUuNjI1IDQyLjE4NzUsMTUuNjI1IHoiCiAgICAg%0D%0AICAgIHN0eWxlPSJmb250LXN0eWxlOm5vcm1hbDtmb250LXZhcmlhbnQ6bm9ybWFsO2ZvbnQtd2Vp%0D%0AZ2h0Om5vcm1hbDtmb250LXN0cmV0Y2g6bm9ybWFsO2ZvbnQtc2l6ZTozMTIuNXB4O2xpbmUtaGVp%0D%0AZ2h0OjEyNSU7Zm9udC1mYW1pbHk6Q29tZm9ydGFhOy1pbmtzY2FwZS1mb250LXNwZWNpZmljYXRp%0D%0Ab246J0NvbWZvcnRhYSwgTm9ybWFsJzt0ZXh0LWFsaWduOnN0YXJ0O3dyaXRpbmctbW9kZTpsci10%0D%0AYjt0ZXh0LWFuY2hvcjpzdGFydCIKICAgICAgICAgaWQ9InBhdGg0MTU0IiAvPgogICAgICA8cGF0%0D%0AaAogICAgICAgICBkPSJtIDk2NC4wMjM5OCw3NTIuMDc3MjEgcSA3LjE4NzUsMi44MTI1IDcuMTg3%0D%0ANSw5LjA2MjUgMCwyLjgxMjUgLTEuMjUsNS42MjUgbCAtMTAxLjI1LDIyMi41IHEgLTIuODEyNSw3%0D%0ALjE4NzUgLTkuMDYyNSw3LjE4NzUgLTMuMTI1LDAgLTUuNjI1LC0xLjI1IC03LjE4NzUsLTIuODEy%0D%0ANSAtNy4xODc1LC05LjM3NSAwLC0yLjgxMjUgMS4yNSwtNS42MjUgbCAzMC45Mzc1LC02OC40Mzc1%0D%0AIDAsLTAuMzEyNSAtNjkuNjg3NSwtMTQ0LjM3NSBxIC0xLjI1LC0yLjUgLTEuMjUsLTUuNjI1IDAs%0D%0ALTYuNTYyNSA2Ljg3NSwtOS4zNzUgMi41LC0xLjI1IDUuNjI1LC0xLjI1IDYuNTYyNSwwIDkuMzc1%0D%0ALDYuNTYyNSBsIDYwLjkzNzUsMTI4LjQzNzUgNTguMTI1LC0xMjguMTI1IHEgMy4xMjUsLTYuODc1%0D%0AIDkuMzc1LC02Ljg3NSAyLjgxMjUsMCA1LjYyNSwxLjI1IHoiCiAgICAgICAgIHN0eWxlPSJmb250%0D%0ALXN0eWxlOm5vcm1hbDtmb250LXZhcmlhbnQ6bm9ybWFsO2ZvbnQtd2VpZ2h0Om5vcm1hbDtmb250%0D%0ALXN0cmV0Y2g6bm9ybWFsO2ZvbnQtc2l6ZTozMTIuNXB4O2xpbmUtaGVpZ2h0OjEyNSU7Zm9udC1m%0D%0AYW1pbHk6Q29tZm9ydGFhOy1pbmtzY2FwZS1mb250LXNwZWNpZmljYXRpb246J0NvbWZvcnRhYSwg%0D%0ATm9ybWFsJzt0ZXh0LWFsaWduOnN0YXJ0O3dyaXRpbmctbW9kZTpsci10Yjt0ZXh0LWFuY2hvcjpz%0D%0AdGFydCIKICAgICAgICAgaWQ9InBhdGg0MTU2IiAvPgogICAgICA8cGF0aAogICAgICAgICBkPSJt%0D%0AIDEwNzcuMDk1Myw4OTguMDE0NzEgcSA1LDAgOC4xMjUsMy40Mzc1IDMuMTI1LDMuNDM3NSAzLjEy%0D%0ANSw4Ljc1IDAsNSAtMy43NSw4LjQzNzUgLTMuNzUsMy40Mzc1IC05LjM3NSwzLjQzNzUgbCAtNi41%0D%0ANjI1LDAgcSAtMTYuNTYyNSwwIC0yOS42ODc1LC03LjgxMjUgLTEzLjEyNSwtOC4xMjUgLTIwLjYy%0D%0ANSwtMjEuNTYyNSAtNy4xODc1LC0xMy43NSAtNy4xODc1LC0zMC45Mzc1IGwgMCwtODMuNDM3NSAt%0D%0AMjAuMzEyNTMsMCBxIC01LDAgLTguMTI1LC0yLjgxMjUgLTIuODEyNSwtMy4xMjUgLTIuODEyNSwt%0D%0ANy41IDAsLTQuNjg3NSAyLjgxMjUsLTcuNSAzLjEyNSwtMy4xMjUgOC4xMjUsLTMuMTI1IGwgMjAu%0D%0AMzEyNTMsMCAwLC00OC40Mzc1IHEgMCwtNS4zMTI1IDMuMTI1LC04Ljc1IDMuNDM3NSwtMy40Mzc1%0D%0AIDguNzUsLTMuNDM3NSA1LjMxMjUsMCA4Ljc1LDMuNDM3NSAzLjQzNzUsMy40Mzc1IDMuNDM3NSw4%0D%0ALjc1IGwgMCw0OC40Mzc1IDM1LjMxMjUsMCBxIDUsMCA3LjgxMjUsMy4xMjUgMy4xMjUsMi44MTI1%0D%0AIDMuMTI1LDcuNSAwLDQuMzc1IC0zLjEyNSw3LjUgLTIuODEyNSwyLjgxMjUgLTcuODEyNSwyLjgx%0D%0AMjUgbCAtMzUuMzEyNSwwIDAsODMuNDM3NSBxIDAsMTUuOTM3NSA5LjM3NSwyNi4yNSA5LjM3NSwx%0D%0AMCAyNC4wNjI1LDEwIGwgOC40Mzc1LDAgeiIKICAgICAgICAgc3R5bGU9ImZvbnQtc3R5bGU6bm9y%0D%0AbWFsO2ZvbnQtdmFyaWFudDpub3JtYWw7Zm9udC13ZWlnaHQ6bm9ybWFsO2ZvbnQtc3RyZXRjaDpu%0D%0Ab3JtYWw7Zm9udC1zaXplOjMxMi41cHg7bGluZS1oZWlnaHQ6MTI1JTtmb250LWZhbWlseTpDb21m%0D%0Ab3J0YWE7LWlua3NjYXBlLWZvbnQtc3BlY2lmaWNhdGlvbjonQ29tZm9ydGFhLCBOb3JtYWwnO3Rl%0D%0AeHQtYWxpZ246c3RhcnQ7d3JpdGluZy1tb2RlOmxyLXRiO3RleHQtYW5jaG9yOnN0YXJ0IgogICAg%0D%0AICAgICBpZD0icGF0aDQxNTgiIC8+CiAgICAgIDxwYXRoCiAgICAgICAgIGQ9Im0gMTIwOC40NTI3%0D%0ALDc0OS4yNjQ3MSBxIDIyLjE4NzUsMCAzOS42ODc1LDkuMzc1IDE3LjgxMjUsOS4zNzUgMjcuODEy%0D%0ANSwyNy41IDEwLDE3LjgxMjUgMTAsNDIuODEyNSBsIDAsODEuMjUgcSAwLDUgLTMuNDM3NSw4LjQz%0D%0ANzUgLTMuNDM3NSwzLjQzNzUgLTguNDM3NSwzLjQzNzUgLTUuMzEyNSwwIC04Ljc1LC0zLjQzNzUg%0D%0ALTMuNDM3NSwtMy40Mzc1IC0zLjQzNzUsLTguNDM3NSBsIDAsLTgxLjI1IHEgMCwtMjcuODEyNSAt%0D%0AMTUuNjI1LC00Mi4xODc1IC0xNS4zMTI1LC0xNC42ODc1IC00MC4zMTI1LC0xNC42ODc1IC0xNS4z%0D%0AMTI1LDAgLTI3LjgxMjUsNS45Mzc1IC0xMi4xODc1LDUuOTM3NSAtMTkuMzc1LDE2LjU2MjUgLTcu%0D%0AMTg3NSwxMC4zMTI1IC03LjE4NzUsMjMuMTI1IGwgMCw5Mi41IHEgMCw1LjMxMjUgLTMuNDM3NSw4%0D%0ALjc1IC0zLjEyNSwzLjEyNSAtOC40Mzc1LDMuMTI1IC01LjMxMjUsMCAtOC43NSwtMy4xMjUgLTMu%0D%0ANDM3NSwtMy40Mzc1IC0zLjQzNzUsLTguNzUgbCAwLC0yMjAgcSAwLC01LjMxMjUgMy40Mzc1LC04%0D%0ALjc1IDMuNDM3NSwtMy40Mzc1IDguNzUsLTMuNDM3NSA1LjMxMjUsMCA4LjQzNzUsMy40Mzc1IDMu%0D%0ANDM3NSwzLjQzNzUgMy40Mzc1LDguNzUgbCAwLDgzLjc1IHEgMTAuMzEyNSwtMTEuNTYyNSAyNSwt%0D%0AMTguMTI1IDE0LjY4NzUsLTYuNTYyNSAzMS44NzUsLTYuNTYyNSB6IgogICAgICAgICBzdHlsZT0i%0D%0AZm9udC1zdHlsZTpub3JtYWw7Zm9udC12YXJpYW50Om5vcm1hbDtmb250LXdlaWdodDpub3JtYWw7%0D%0AZm9udC1zdHJldGNoOm5vcm1hbDtmb250LXNpemU6MzEyLjVweDtsaW5lLWhlaWdodDoxMjUlO2Zv%0D%0AbnQtZmFtaWx5OkNvbWZvcnRhYTstaW5rc2NhcGUtZm9udC1zcGVjaWZpY2F0aW9uOidDb21mb3J0%0D%0AYWEsIE5vcm1hbCc7dGV4dC1hbGlnbjpzdGFydDt3cml0aW5nLW1vZGU6bHItdGI7dGV4dC1hbmNo%0D%0Ab3I6c3RhcnQiCiAgICAgICAgIGlkPSJwYXRoNDE2MCIgLz4KICAgICAgPHBhdGgKICAgICAgICAg%0D%0AZD0ibSAxNDE3LjExOTcsOTIzLjYzOTcxIHEgLTI0LjY4NzUsMCAtNDQuNjg3NSwtMTEuMjUgLTE5%0D%0ALjY4NzUsLTExLjI1IC0zMC45Mzc1LC0zMC45Mzc1IC0xMS4yNSwtMjAgLTExLjI1LC00NC42ODc1%0D%0AIDAsLTI0LjY4NzUgMTEuMjUsLTQ0LjY4NzUgMTEuMjUsLTIwIDMwLjkzNzUsLTMxLjI1IDIwLC0x%0D%0AMS4yNSA0NC42ODc1LC0xMS4yNSAyNC42ODc1LDAgNDQuMzc1LDExLjI1IDE5LjY4NzUsMTEuMjUg%0D%0AMzAuOTM3NSwzMS4yNSAxMS41NjI1LDIwIDExLjU2MjUsNDQuNjg3NSAwLDI0LjY4NzUgLTExLjI1%0D%0ALDQ0LjY4NzUgLTExLjI1LDE5LjY4NzUgLTMxLjI1LDMwLjkzNzUgLTE5LjY4NzUsMTEuMjUgLTQ0%0D%0ALjM3NSwxMS4yNSB6IG0gMCwtMjEuODc1IHEgMTguMTI1LDAgMzIuNSwtOC40Mzc1IDE0LjM3NSwt%0D%0AOC40Mzc1IDIyLjUsLTIzLjEyNSA4LjEyNSwtMTUgOC4xMjUsLTMzLjQzNzUgMCwtMTguNzUgLTgu%0D%0AMTI1LC0zMy40Mzc1IC04LjEyNSwtMTUgLTIyLjUsLTIzLjQzNzUgLTE0LjM3NSwtOC40Mzc1IC0z%0D%0AMi41LC04LjQzNzUgLTE3LjgxMjUsMCAtMzIuNSw4LjQzNzUgLTE0LjM3NSw4LjQzNzUgLTIyLjgx%0D%0AMjUsMjMuNDM3NSAtOC4xMjUsMTQuNjg3NSAtOC4xMjUsMzMuNDM3NSAwLDE4LjQzNzUgOC4xMjUs%0D%0AMzMuNDM3NSA4LjQzNzUsMTQuNjg3NSAyMi44MTI1LDIzLjEyNSAxNC42ODc1LDguNDM3NSAzMi41%0D%0ALDguNDM3NSB6IgogICAgICAgICBzdHlsZT0iZm9udC1zdHlsZTpub3JtYWw7Zm9udC12YXJpYW50%0D%0AOm5vcm1hbDtmb250LXdlaWdodDpub3JtYWw7Zm9udC1zdHJldGNoOm5vcm1hbDtmb250LXNpemU6%0D%0AMzEyLjVweDtsaW5lLWhlaWdodDoxMjUlO2ZvbnQtZmFtaWx5OkNvbWZvcnRhYTstaW5rc2NhcGUt%0D%0AZm9udC1zcGVjaWZpY2F0aW9uOidDb21mb3J0YWEsIE5vcm1hbCc7dGV4dC1hbGlnbjpzdGFydDt3%0D%0Acml0aW5nLW1vZGU6bHItdGI7dGV4dC1hbmNob3I6c3RhcnQiCiAgICAgICAgIGlkPSJwYXRoNDE2%0D%0AMiIgLz4KICAgICAgPHBhdGgKICAgICAgICAgZD0ibSAxNjMyLjAzNjcsNzQ5LjI2NDcxIHEgMjIu%0D%0AMTg3NSwwIDM5LjY4NzUsOS4zNzUgMTcuODEyNSw5LjM3NSAyNy44MTI1LDI3LjUgMTAsMTcuODEy%0D%0ANSAxMCw0Mi44MTI1IGwgMCw4MS4yNSBxIDAsNSAtMy40Mzc1LDguNDM3NSAtMy40Mzc1LDMuNDM3%0D%0ANSAtOC40Mzc1LDMuNDM3NSAtNS4zMTI1LDAgLTguNzUsLTMuNDM3NSAtMy40Mzc1LC0zLjQzNzUg%0D%0ALTMuNDM3NSwtOC40Mzc1IGwgMCwtODEuMjUgcSAwLC0yNy44MTI1IC0xNS42MjUsLTQyLjE4NzUg%0D%0ALTE1LjMxMjUsLTE0LjY4NzUgLTQwLjMxMjUsLTE0LjY4NzUgLTE1LjMxMjUsMCAtMjcuODEyNSw1%0D%0ALjkzNzUgLTEyLjE4NzUsNS45Mzc1IC0xOS4zNzUsMTYuNTYyNSAtNy4xODc1LDEwLjMxMjUgLTcu%0D%0AMTg3NSwyMy4xMjUgbCAwLDkyLjUgcSAwLDUuMzEyNSAtMy40Mzc1LDguNzUgLTMuMTI1LDMuMTI1%0D%0AIC04LjQzNzUsMy4xMjUgLTUuMzEyNSwwIC04Ljc1LC0zLjEyNSAtMy40Mzc1LC0zLjQzNzUgLTMu%0D%0ANDM3NSwtOC43NSBsIDAsLTE0Ni44NzUgcSAwLC01LjMxMjUgMy40Mzc1LC04Ljc1IDMuNDM3NSwt%0D%0AMy40Mzc1IDguNzUsLTMuNDM3NSA1LjMxMjUsMCA4LjQzNzUsMy40Mzc1IDMuNDM3NSwzLjQzNzUg%0D%0AMy40Mzc1LDguNzUgbCAwLDEwLjYyNSBxIDEwLjMxMjUsLTExLjU2MjUgMjUsLTE4LjEyNSAxNC42%0D%0AODc1LC02LjU2MjUgMzEuODc1LC02LjU2MjUgeiIKICAgICAgICAgc3R5bGU9ImZvbnQtc3R5bGU6%0D%0Abm9ybWFsO2ZvbnQtdmFyaWFudDpub3JtYWw7Zm9udC13ZWlnaHQ6bm9ybWFsO2ZvbnQtc3RyZXRj%0D%0AaDpub3JtYWw7Zm9udC1zaXplOjMxMi41cHg7bGluZS1oZWlnaHQ6MTI1JTtmb250LWZhbWlseTpD%0D%0Ab21mb3J0YWE7LWlua3NjYXBlLWZvbnQtc3BlY2lmaWNhdGlvbjonQ29tZm9ydGFhLCBOb3JtYWwn%0D%0AO3RleHQtYWxpZ246c3RhcnQ7d3JpdGluZy1tb2RlOmxyLXRiO3RleHQtYW5jaG9yOnN0YXJ0Igog%0D%0AICAgICAgICBpZD0icGF0aDQxNjQiIC8+CiAgICA8L2c+CiAgICA8ZwogICAgICAgaWQ9Imc0MzU5%0D%0AIgogICAgICAgdHJhbnNmb3JtPSJ0cmFuc2xhdGUoMiw2KSI+CiAgICAgIDxnCiAgICAgICAgIGlk%0D%0APSJnNDQ3NyI+CiAgICAgICAgPHBhdGgKICAgICAgICAgICBpbmtzY2FwZTpjb25uZWN0b3ItY3Vy%0D%0AdmF0dXJlPSIwIgogICAgICAgICAgIGlkPSJwYXRoNDE5NCIKICAgICAgICAgICBkPSJtIDQ0NC42%0D%0AMzcxNiwtNTYuMzk2MTg0IGMgLTE0Ljg4MDMxLC0xLjI5NDQ2NCAtMjguNDY2ODcsMS4yOTQ0NjIg%0D%0ALTQwLjQzNTUzLDYuMTQ2MTczIC0yMy4yOTExNCw5LjA1NzU2OCAtNDAuNzU5NDMsMjMuMjkxIC01%0D%0ANC45OTI2MSwzNi41NTM5NjYgLTIyLjk2NzYzLDIxLjk5Njk5NCAtMzguNDk1MDQsNDYuNTgxNjkx%0D%0AIC01Mi40MDUwOSw2OC4yNTU0NTcgLTE5LjQwODgzLDMwLjczMTMzIC0zMi42NzE4MSw2MS43ODU3%0D%0AMjggLTQ2LjI1ODMxLDg5LjI4MjI4OCBsIDAsMCAwLDAgYyAtMTcuNDY3OSwzNS4yNTk4IC0zMi45%0D%0AOTUzNSw3MC41MTk2MSAtNDguODQ2MzEsMTA0LjE2MjM0IGwgMCwwIDAsMCBjIDAsMCAtMjEuMzUw%0D%0AMjIsNDQuOTY0NDQgLTMwLjA4Mzk2LDYzLjcyNjY2IC0xMS42NDU1NiwyNS41NTUxNiAtMjIuOTY3%0D%0ANzUsNTIuNDA0NzkgLTMyLjY3MjQsNzEuNDg5MTYgLTkuNzA0NjQsMTkuMDg1OTEgLTEyLjI5MjUs%0D%0AMjUuMjMyMDggLTI1Ljg3ODU0LDQ2LjU4MjMgLTEzLjU4NjUwNCwyMS4zNTAyMyAtNDguMDUwMjU0%0D%0ALDkuMDgwODggLTU4Ljg0NTk4Niw0NC45ODY1MSAtMjQuMTU0NTM3LDgwLjMzNTgxIC0yNy40MjA0%0D%0AOTEsMTAwLjYyNDc1IDIuMDk4NDMzLDExOS42MDkyNSAzMy44NDcyMiwyMS4yMjA4OCA1MC44MDA4%0D%0AMDMsMTMuNzgyODkgMTE1LjYyMTg1MywtNDkuMjY4OTIgMjYuODQ1MiwtMjYuMTEyNSAwLjk3MDUs%0D%0ALTQ4LjY4ODE3IDEzLjkwOTYyLC02OS4wNjc5MiAxMi45Mzk1OCwtMjAuMzc5NzYgMTguNzYyMzcs%0D%0ALTMyLjAyNTMzIDMwLjA4NDQsLTUzLjY5OTA5IDEwLjk5ODY3LC0yMS42NzM3NyAyMy4yOTA3LC01%0D%0AMS40MzMyNSAzMy45NjU4LC03NC40MDEwMSA4LjQxMDgyLC0xNy43OTE3NSAyOS43NjEwNCwtNjMu%0D%0AMDc5NzUgMjkuNzYxMDQsLTYzLjA3OTc1IGwgMCwwIGMgMTYuNDk3OSwtMzUuMjU5ODEgMzIuMDI0%0D%0AODYsLTY5Ljg3MyA0OC4xOTkyMiwtMTAzLjE5MTg3IDE0LjU1Njk4LC0yOS43NjA4NyAyNy4xNzI1%0D%0ANiwtNTcuNTgwNSA0MS43Mjk1NywtODAuODcxNjUgMTMuMjYzMTIsLTIxLjAyNjIyIDI1Ljg3OTEx%0D%0ALC0zOS4xNDE2NjIgMzkuMTQyMTIsLTUxLjc1NzcwMiA5LjcwNDEyLC05LjM4MTEwOCAxOS40MDg4%0D%0AMiwtMTYuMTc0MzYyIDI3LjE3MjUzLC0xOS4wODUyOTUgMC42NDY0MSwtMC4zMjM5OTggMS42MTc1%0D%0ANSwtMC4zMjM5OTggMC42NDY0MSwtMC4zMjM5OTggNy4xMTY4NCwwLjY0NjQ2MSAxNS4yMDQxMSwz%0D%0ALjU1ODAxIDE5LjczMjg0LDYuNDY5NDA4IDE4LjQzNzY4LDEyLjkzOTQyIDM1LjI1ODg4LDM4LjQ5%0D%0ANTA0MiA1NS4zMTUyNiw2OC4yNTU0NDcgMTkuNzMyNzMsMjkuNzYxMDIgMzUuMjYwMTUsNjYuNjM4%0D%0AMjIgNTQuMzQ1OTksMTA1Ljc3OTg4IDI3LjQ5NjQ0LDU2LjYxMDA0IDUyLjQwNTEzLDExNi4xMzE0%0D%0ANSA4MC44NzE5OSwxNzUuMzI4NDIgMTIuNjE2MDEsMjYuODQ5NDcgMjUuMjMyMTMsNTUuOTYzNDIg%0D%0ANDIuMDUzNTQsODUuNDAwODkgMTAuOTk4NywxOS40MDkzIDIzLjYxNDcsMzguMTcxNjYgMzYuODc3%0D%0ANyw1NS45NjM0MSA3LjExNjcxLDkuNzA0NjQgMTYuODIxNDIsMjcuMTczIDM3LjUyNDU3LDQyLjcw%0D%0AMDQ0IDE3LjE0NDk3LDEyLjkzOTU4IDQzLjAyMzk4LDM0LjYxMzM0IDgzLjEzNjUyLDMyLjk5NTgg%0D%0ANDIuMzc2ODcsLTEuNjE3NCA2OC41NzUyNSwtMjcuMTczMDIgODYuMzY2OTgsLTQzLjAyMzgzIDMx%0D%0ALjM3ODQ2LC0yNy44MjAwOCA0Ni4yNTg4NCwtNjAuNDkyMzUgNjAuODE1ODMsLTgzLjQ1OTk3IDEw%0D%0ALjk5ODU2LC0xNy43OTE5IDIxLjAyNjc2LC0zNS45MDcxNyAyOS43NjA4NiwtNTMuNjk5MDkgMTMu%0D%0AMjYzLC0yNS44NzkgMjMuNjE0NywtNTEuMTA5ODUgMzQuMjg5OSwtNzQuMDc3NDYgNy43NjM3LC0x%0D%0ANi44MjE0NCAyNS4yMzIxLC01My42OTg2NCAyNy40OTY0LC01OS4xOTc5IGwgMC4zMjM4LDAgMS45%0D%0ANDExLC00LjUyODkzIGMgMCwwIDMyLjAyNTMsLTY5Ljg3MyA0OC4xOTk4LC0xMDIuNTQ0NzkgMTQu%0D%0ANTU2OSwtMjkuNzYxMDIgMjYuODQ5NCwtNTcuOTA0MDMgNDEuNzMsLTgxLjE5NTE5IDEzLjI2Mywt%0D%0AMjAuNzAyODMyIDI1LjU1NTQsLTM5LjE0MTY2OCAzOS4xNDE5LC01MS43NTc3MDcgOS43MDQ3LC05%0D%0ALjM4MTEwOSAxOS40MDkzLC0xNi4xNzQzNSAyNy4xNzMxLC0xOS4wODUyOTYgMC42NDYzLC0wLjMy%0D%0AMzk5OCAxLjYxNzUsLTAuMzIzOTk4IDAuNjQ2MywtMC4zMjM5OTggNy4xMTY5LDAuNjQ2NDc0IDE0%0D%0ALjg4MDYsMy41NTgwMjMgMTkuNDA5NCw2Ljc5MjgwMiAxOC40Mzg4LDEyLjYxNjAzOSAzNS41ODM4%0D%0ALDM4LjE3MTY2MiA1NS4zMTY1LDY3LjkzMjA2OSAyMC4wNTYzLDI5Ljc2MTAxIDM1LjI2MDMsNjYu%0D%0AOTYxNzUgNTQuMzQ2LDEwNS43Nzk4OSAyNy44Miw1Ni42MTAwMSA1Mi43Mjg3LDExNi4xMzE0NCA4%0D%0AMS4xOTU3LDE3NS42NTE5NCAzMi42NzIyLDc5LjU3ODEgMTExLjkyMjUsMTgwLjgyOTggMTQxLjM2%0D%0AMDEsMTg1LjM1ODc0IDYuNzkzMSwtMjkuNzYxMDEgLTI3LjQ5NjQsLTE0Ni41NDAxNSAtNjMuNzI3%0D%0ANCwtMjIyLjU1ODU2IC0yNy4xNzMsLTU2LjkzMzU2IC01Mi4wNzcyLC0xMTYuNzc4MzggLTgxLjE5%0D%0AMTEsLTE3Ni4yOTk3OSAtMTcuNDY4NSwtMzUuNTgzMzUgLTMzLjk2NjMsLTc2LjM0Mjg3MSAtNjAu%0D%0AMTY4OSwtMTE1LjgwNzkxOCAtMTguNDM4OSwtMjcuODE5NjMgLTM5LjQ2NTUsLTY0LjM3Mzc0OTkg%0D%0ALTc3Ljk2MDYsLTkwLjg5OTUzIC0xOC43NjIzLC0xMy4yNjI5NjUgLTM5Ljc4OTEsLTIwLjA1NjIx%0D%0ANyAtNjAuODE1OSwtMjEuOTk2OTkzIC0xNC41NTY5LC0xLjI5NDQ2MyAtMjguMTQzNCwxLjI5NDQ2%0D%0AMyAtNDAuMTEyNCw2LjE0NjE3NCAtMjMuMjkxMiw5LjA1NzU2OCAtNDAuNzU5NCwyMy4yOTA5OTgg%0D%0ALTU0Ljk5MzIsMzYuODc3MTk4IC0yMi45Njc1LDIxLjY3Mzc2MTIgLTM4LjgxODUsNDYuNTgyMDAx%0D%0AIC01Mi40MDUsNjcuOTMyMjIyIC0xOS40MDkyLDMwLjczMTMzIC0zMi45OTU4LDYxLjc4NTczNyAt%0D%0ANDYuNTgyNCw4OS4yODIyODcgLTE2LjgyMTEsMzQuMjg5MzQgLTQzLjk5NDIsOTQuNDU3NjkgLTQ2%0D%0ALjU4MjIsMTAwLjI4MDQ4IC0wLjMyNDEsMCAtMC4zMjQxLDAuNjQ2NDYgLTAuNjQ2NSwwLjY0NjQ2%0D%0AIGwgLTEuNjE3MywzLjU1ODMyIGMgMCwwIC0yMS4wMjY3OCw0NC42NDA5MiAtMjkuNzYwODgsNjMu%0D%0ANDAzMyAtMTEuNjQ1NTksMjUuNTU1MTUgLTIxLjY3Mzc0LDQ5LjgxNjc3IC0zMi42NzI0Niw3MS44%0D%0AMTI2OCAtOC40MTA1OCwxNi4xNzQzNyAtMTYuODIxMywzMi4wMjUzNCAtMjYuMjAyNTYsNDYuMjU4%0D%0ANzcgLTE3Ljc4NzQzLDI4LjQ2NzAxIC0zMC43MjY5OSw1Mi40MDUwOSAtNDQuOTYwNDIsNjUuMDIx%0D%0AMTMgLTE1Ljg1MDk5LDE0LjIzMzQzIC0zMC4wODQ0MywyMS4zNTAyMiAtMzIuNjcyMjksMjEuNjcz%0D%0ANzYgLTAuMzI0MDIsMCAtMTIuNjE2LC00Ljg1MjMyIC0yOC4xNDM1MywtMTYuMTc0NTEgLTIuOTEx%0D%0AMzEsLTEuOTQwOTMgLTkuMzgxMjYsLTEwLjk5ODUgLTIwLjA1NjI3LC0yNC45MDg1NCBDIDc4My42%0D%0ANTE2LDQ2MC4yMDg3MyA3NzMuMjk5OTEsNDQ0LjM1Nzc2IDc2My45MTg3Nyw0MjguMTgzNDEgNzUw%0D%0ALjMzMjMzLDQwNC4yNDUxNyA3MzguMzYzMjEsMzc3LjA3MjE2IDcyNC43NzY2NSwzNDguNjA2NTMg%0D%0ANjk3LjYwMzY2LDI5MS4zNDk1NiA2NzIuMzcxNjQsMjMxLjgyODE1IDY0My4yNTc2NywxNzIuMzA2%0D%0ANzIgNjI1Ljc4OTI2LDEzNi43MjM1MiA2MDkuMjkxNCw5NS45NjQwMTYgNTgzLjQxMjM5LDU2LjQ5%0D%0AODgwNiA1NjQuNjQ5OTksMjguNjc5MzI3IDU0My42MjMzOSwtNy44NzQ3ODU4IDUwNS40NTE3MSwt%0D%0AMzQuNzI0MTA1IDQ4Ni42ODk0MiwtNDcuNjYzNTMxIDQ2NS42NjM5OSwtNTQuNDU2NjMxIDQ0NC42%0D%0AMzcxNiwtNTYuMzk3NTYxIFoiCiAgICAgICAgICAgc3R5bGU9ImZpbGw6IzM3NzA5ZjtmaWxsLW9w%0D%0AYWNpdHk6MTtmaWxsLXJ1bGU6bm9uemVybztzdHJva2U6bm9uZSIKICAgICAgICAgICBzb2RpcG9k%0D%0AaTpub2RldHlwZXM9ImNjY2NjY2NjY2NjY2NzY3NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2Nj%0D%0AY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjIiAvPgogICAgICAgIDxwYXRoCiAgICAg%0D%0AICAgICAgaW5rc2NhcGU6Y29ubmVjdG9yLWN1cnZhdHVyZT0iMCIKICAgICAgICAgICBpZD0icGF0%0D%0AaDQxOTYiCiAgICAgICAgICAgZD0ibSA3Ni4yMjk0MzgsNTI5LjQ4MDcyIGMgMCwzLjU1ODMyIC0y%0D%0ALjkxMTQwMyw2LjQ2OTcxIC02LjQ2OTcyNiw2LjQ2OTcxIC0zLjU1ODMxMiwwIC02LjQ2OTcwOSwt%0D%0AMi45MTEzOSAtNi40Njk3MDksLTYuNDY5NzEgMCwtMy41NTgzMyAyLjkxMTM5NywtNi40Njk3MSA2%0D%0ALjQ2OTcwOSwtNi40Njk3MSAzLjU1ODMyMywwIDYuNDY5NzI2LDIuOTExMzggNi40Njk3MjYsNi40%0D%0ANjk3MSB6IgogICAgICAgICAgIHN0eWxlPSJmaWxsOiNmZmZmZmY7ZmlsbC1vcGFjaXR5OjE7Zmls%0D%0AbC1ydWxlOm5vbnplcm87c3Ryb2tlOm5vbmUiIC8+CiAgICAgICAgPHBhdGgKICAgICAgICAgICBp%0D%0AbmtzY2FwZTpjb25uZWN0b3ItY3VydmF0dXJlPSIwIgogICAgICAgICAgIGlkPSJwYXRoNDE5OCIK%0D%0AICAgICAgICAgICBkPSJtIDExNS44OTM3LDU1NS4yNzYyMyBjIDAsMy41NTgzIC0yLjU4Nzg1LDYu%0D%0ANDY5NzEgLTYuNDY5Nyw2LjQ2OTcxIC0zLjU1ODMyLDAgLTYuNDY5NzMsLTIuOTExNDEgLTYuNDY5%0D%0ANzMsLTYuNDY5NzEgMCwtMy41NTg0OCAyLjkxMTQxLC02LjQ2OTg4IDYuNDY5NzMsLTYuNDY5ODgg%0D%0AMy44ODE4NSwwIDYuNDY5NywyLjkxMTQgNi40Njk3LDYuNDY5ODggeiIKICAgICAgICAgICBzdHls%0D%0AZT0iZmlsbDojZmZmZmZmO2ZpbGwtb3BhY2l0eToxO2ZpbGwtcnVsZTpub256ZXJvO3N0cm9rZTpu%0D%0Ab25lIiAvPgogICAgICA8L2c+CiAgICAgIDxnCiAgICAgICAgIGlkPSJnNDM0NSIKICAgICAgICAg%0D%0AdHJhbnNmb3JtPSJtYXRyaXgoLTAuOTcyMDIxMTIsMCwwLC0wLjk3MjAyMTEyLDE4MTMuOTA4NCw1%0D%0AMjcuMDUyMjIpIj4KICAgICAgICA8cGF0aAogICAgICAgICAgIHN0eWxlPSJmaWxsOiNmZmQxNDE7%0D%0AZmlsbC1vcGFjaXR5OjE7ZmlsbC1ydWxlOm5vbnplcm87c3Ryb2tlOm5vbmUiCiAgICAgICAgICAg%0D%0AZD0ibSAxNzExLjA2NDUsODcuMzc1IGMgLTE5Ljk3NzEsMC4zMjgzMDIgLTQxLjkxOTIsMTcuNTg2%0D%0AMjggLTg1LjQ3MDcsNTkuOTQ5MjIgLTI2Ljg0NTMsMjYuMTEyNDkgLTAuOTcxMSw0OC42ODg2MSAt%0D%0AMTMuOTEwMiw2OS4wNjgzNiAtMTIuOTM5NiwyMC4zNzk3NiAtMTguNzYyLDMyLjAyNTQ2IC0zMC4w%0D%0AODQsNTMuNjk5MjIgLTEwLjk5ODcsMjEuNjczNzYgLTIzLjI5MTcsNTEuNDMyNjMgLTMzLjk2Njgs%0D%0ANzQuNDAwMzkgLTguNDEwOCwxNy43OTE3NSAtMjkuNzU5OCw2My4wODAwOCAtMjkuNzU5OCw2My4w%0D%0AODAwOCAtNC4xNTQ4LDguODc5NzYgLTguMjQ3MiwxNy43MTY1NSAtMTIuMzA0NiwyNi41MDM5IDku%0D%0ANDQ5NCwyMC45OTM3MSAxOC44OTA3LDQxLjc3MjYxIDI4LjYyNSw2Mi4xNjc5NyA2LjE0NDEsMTIu%0D%0AODkxMjkgMTIuMjI2NywyNi45Njc1NiAxOC4wNjgzLDQxLjU4MDA4IDE1LjM1MTYsLTMxLjY0Nzc3%0D%0AIDI5LjMzNjQsLTYzLjE2NDIgNDMuNTcwMywtOTMuMzc1IDAsMCAyMS4zNTAzLC00NC45NjQzNSAz%0D%0AMC4wODQsLTYzLjcyNjU2IDExLjY0NTYsLTI1LjU1NTE3IDIyLjk2OTIsLTUyLjQwMzkyIDMyLjY3%0D%0AMzgsLTcxLjQ4ODI4IDkuNzA0NywtMTkuMDg1OTIgMTIuMjkxLC0yNS4yMzE4MiAyNS44NzcsLTQ2%0D%0ALjU4MjA0IDEzLjU4NjUsLTIxLjM1MDIzIDQ4LjA1LC05LjA4MjY0IDU4Ljg0NTcsLTQ0Ljk4ODI4%0D%0AIDI0LjE1NDUsLTgwLjMzNTgxIDI3LjQyMTMsLTEwMC42MjQ4OCAtMi4wOTc3LC0xMTkuNjA5Mzcy%0D%0AIC0xMS4xMDYxLC02Ljk2MzEwMyAtMjAuMzk0MSwtMTAuODQwMDIyIC0zMC4xNTAzLC0xMC42Nzk2%0D%0AODggeiBtIC03NTYuMjIyNywyLjQ0OTIxOSBjIC0xLjIyNTg5LDAuMDA2IC0yLjQ2NTI0LDAuMDMz%0D%0ANDQgLTMuNzE4NzUsMC4wODM5OCAtNDIuMzc2ODcsMS42MTczOTUgLTY4LjU3MzUxLDI3LjE3NDU4%0D%0AMSAtODYuMzY1MjQsNDMuMDI1MzkxIC0zMS4zNzg0NiwyNy44MjAwOCAtNDYuMjU5NDIsNjAuNDkx%0D%0AMzYgLTYwLjgxNjQsODMuNDU4OTkgLTEwLjk5ODUyLDE3Ljc5MTg5IC0yMS4wMjc2MywzNS45MDcz%0D%0AIC0yOS43NjE3Miw1My42OTkyMiAtMTMuMjYzMDQsMjUuODc5IC0yMy42MTM4NCw1MS4xMTA1MSAt%0D%0AMzQuMjg5MDcsNzQuMDc4MTIgLTcuNzYzNzEsMTYuODIxNDQgLTI1LjIzMTc3LDUzLjY5ODAxIC0y%0D%0ANy40OTYwOSw1OS4xOTcyNyBsIC0wLjMyNDIyLDAgLTEuOTQxNCw0LjUyOTI5IGMgMCwwIC01LjM0%0D%0ANTksMTEuNjE4MTYgLTEyLjA0ODgzLDI2LjA5NTcxIDkuNDk1MDQsMjAuOTY4NSAxOC45NzIwMSw0%0D%0AMS43NTk2IDI4LjY5NzI2LDYyLjI1MTk1IDYuNTI2NTQsMTMuNjczOTYgMTIuNjc4MTMsMjcuMDM2%0D%0ANDkgMTguODE2NDEsMzkuODk4NDQgMTYuNjA3LC0zNC43MDI4MSAzOC4yNTg1MywtODIuNjQgNDAu%0D%0ANTU2NjQsLTg3LjgxMDU1IDAuMzI0MDcsMCAwLjMyNDA3LC0wLjY0NjQ4IDAuNjQ2NDksLTAuNjQ2%0D%0ANDggbCAxLjYxNzE4LC0zLjU1ODYgYyAwLDAgMjEuMDI1NjcsLTQ0LjYzOTk2IDI5Ljc1OTc3LC02%0D%0AMy40MDIzNCAxMS42NDU1OSwtMjUuNTU1MTYgMjEuNjczMTUsLTQ5LjgxNjU5IDMyLjY3MTg3LC03%0D%0AMS44MTI1IDguNDEwNTksLTE2LjE3NDM2IDE2LjgyMTg3LC0zMi4wMjYzMyAyNi4yMDMxMywtNDYu%0D%0AMjU5NzcgMTcuNzg3NDMsLTI4LjQ2NzAxIDMwLjcyNzUsLTUyLjQwNTQ0IDQ0Ljk2MDk0LC02NS4w%0D%0AMjE0OCAxNS44NTA5OCwtMTQuMjMzNDMgMzAuMDg0MDEsLTIxLjM1MDI4IDMyLjY3MTg3LC0yMS42%0D%0ANzM4MyAwLjMyNDAyLDAgMTIuNjE3LDQuODUzNTkgMjguMTQ0NTMsMTYuMTc1NzggMi45MTEzMSwx%0D%0ALjk0MDkyIDkuMzc5NjMsMTAuOTk4MTcgMjAuMDU0NzMsMjQuOTA4MjEgMTAuOTk4NSwxNS4yMDM4%0D%0AOSAyMS4zNTEyLDMxLjA1NDE2IDMwLjczMjQsNDcuMjI4NTEgMTMuNTg2NCwyMy45MzgyNCAyNS41%0D%0ANTYsNTEuMTEyNDkgMzkuMTQyNiw3OS41NzgxMyAyNy4xNzMsNTcuMjU2OTUgNTIuNDAzNiwxMTYu%0D%0ANzc3NCA4MS41MTc1LDE3Ni4yOTg4MiAxNy40Njg1LDM1LjU4MzIgMzMuOTY2Nyw3Ni4zNDMzOSA1%0D%0AOS44NDU3LDExNS44MDg2IDE4Ljc2MjQsMjcuODE5NDggMzkuNzg5Myw2NC4zNzMzNSA3Ny45NjEs%0D%0AOTEuMjIyNjUgMTguNzYyMywxMi45Mzk0MyAzOS43ODc2LDE5LjczMjkgNjAuODE0NCwyMS42NzM4%0D%0AMyBsIDAsLTAuMDAyIGMgMTQuODgwMywxLjI5NDQ2IDI4LjQ2NjksLTEuMjk0NzcgNDAuNDM1Niwt%0D%0ANi4xNDY0OSAyMy4yOTExLC05LjA1NzU1IDQwLjc1OSwtMjMuMjg5NzYgNTQuOTkyMiwtMzYuNTUy%0D%0ANzMgMjIuOTY3NiwtMjEuOTk2OTkgMzguNDk2MiwtNDYuNTgyMDkgNTIuNDA2MiwtNjguMjU1ODYg%0D%0AMi41MDg0LC0zLjk3MTYyIDQuOTA5NiwtNy45NDc4MSA3LjIyNjYsLTExLjkyMTg3IC0xOS43NzEx%0D%0ALC0yOC45NTE4OSAtMzguMTI0LC02MC45MzgwOCAtNTAuNDAwNCwtOTAuMTkxNDEgLTkuODE5OCwy%0D%0AMC4yODgzMyAtMTkuMjI1NSwzOS4yMzgxNiAtMjkuNjExMyw1NS44NTU0NyAtMTMuMjYzMiwyMS4w%0D%0AMjYyMiAtMjUuODc5NiwzOS4xMzk4MiAtMzkuMTQyNiw1MS43NTU4NiAtOS43MDQxLDkuMzgxMTEg%0D%0ALTE5LjQwODIsMTYuMTc1IC0yNy4xNzE5LDE5LjA4NTk0IC0wLjMzOTUsMC4xNzAxNyAtMC43NTEy%0D%0ALDAuMjQ4NzYgLTAuOTQxNCwwLjI4NzEgLTcuMDQ1MSwtMC42OTgxMSAtMTQuOTcwNSwtMy41NTk5%0D%0AMSAtMTkuNDM3NSwtNi40MzE2NCAtMTguNDM3NywtMTIuOTM5NDIgLTM1LjI2LC0zOC40OTU0NSAt%0D%0ANTUuMzE2NCwtNjguMjU1ODYgLTE5LjczMjcsLTI5Ljc2MSAtMzUuMjU5OSwtNjYuNjM3NjIgLTU0%0D%0ALjM0NTcsLTEwNS43NzkyOSAtMjcuNDk2NCwtNTYuNjEwMDQgLTUyLjQwNDIsLTExNi4xMzMxMSAt%0D%0AODAuODcxMSwtMTc1LjMzMDA4IC0xMi42MTYsLTI2Ljg0OTQ3IC0yNS4yMzMzLC01NS45NjI5MyAt%0D%0ANDIuMDU0NywtODUuNDAwMzkgLTEwLjk5ODcsLTE5LjQwOTMgLTIzLjYxMzksLTM4LjE3MTE0IC0z%0D%0ANi44NzY5LC01NS45NjI4OSAtNy4xMTY4LC05LjcwNDY0IC0xNi44MjIzLC0yNy4xNzM3NCAtMzcu%0D%0ANTI1NCwtNDIuNzAxMTcgLTE2LjYwOTMsLTEyLjUzNTIzIC00MS40MTU1NiwtMzMuMjY3MTQyIC03%0D%0AOS40MTgsLTMzLjA4MDA4NSB6IE0gMjAxLjYwOTM4LDEyMS4yODcxMSBjIC02Ljc5MzA3LDI5Ljc2%0D%0AMTAxIDI3LjQ5NzQ5LDE0Ni41NDAxOSA2My43Mjg1MSwyMjIuNTU4NTkgMjcuMTcyOTUsNTYuOTMz%0D%0ANTYgNTIuMDc3NTMsMTE2Ljc3OTM4IDgxLjE5MTQxLDE3Ni4zMDA3OCAxNy40Njg0OCwzNS41ODMz%0D%0ANiAzMy45NjUzOSw3Ni4zNDM1NSA2MC4xNjc5NywxMTUuODA4NiAxOC40Mzg4NSwyNy44MTk2MyAz%0D%0AOS40NjU4OCw2NC4zNzI2NiA3Ny45NjA5Myw5MC44OTg0NCAxOC43NjIzNCwxMy4yNjI5NiAzOS43%0D%0AODk1NSwyMC4wNTcyNyA2MC44MTY0MSwyMS45OTgwNCAxNC41NTY4OSwxLjI5NDQ3IDI4LjE0MjM1%0D%0ALC0xLjI5NDc3IDQwLjExMTMzLC02LjE0NjQ4IDIzLjI5MTE4LC05LjA1NzU3IDQwLjc2MDM1LC0y%0D%0AMy4yOTA3NiA1NC45OTQxNCwtMzYuODc2OTYgMjIuOTY3NSwtMjEuNjczNzUgMzguODE3ODcsLTQ2%0D%0ALjU4MzM3IDUyLjQwNDMsLTY3LjkzMzU5IDEuNjI1MDYsLTIuNTczMDQgMy4xOTUzMiwtNS4xNDcz%0D%0ANyA0Ljc0MjE4LC03LjcyMjY1IC0yLjM0MDcxLC0zLjg1NzUxIC00LjY0MDE2LC03LjczNTY2IC02%0D%0ALjg1NzQyLC0xMS42NDg0NCAtMTYuMzM5MzcsLTI4LjU5MzkxIC0yOC43MDksLTU2Ljg2NTY1IC00%0D%0AMC45NjY4LC04My4wNzAzMiAtOS43NTY2NCwyMC40MDkyOCAtMTkuMDczOTYsMzkuNTQ2NzggLTI5%0D%0ALjcwMzEyLDU2LjE4MzYgLTEzLjI2MzA0LDIwLjcwMjgzIC0yNS41NTYwNiwzOS4xNDE3NyAtMzku%0D%0AMTQyNTgsNTEuNzU3ODEgLTkuNzA0NjYsOS4zODExMSAtMTkuNDA4MDUsMTYuMTczMDQgLTI3LjE3%0D%0AMTg3LDE5LjA4Mzk5IC0wLjMzMTMxLDAuMTY2MDggLTAuNzM4OTEsMC4yNDc2NiAtMC45MzU1NSww%0D%0ALjI4NzEgLTcuMDM3NzQsLTAuNjk3NDQgLTE0LjY1MzE2LC0zLjU2NDUxIC0xOS4xMjExLC02Ljc1%0D%0ANTg1IC0xOC40Mzg4NCwtMTIuNjE2MDQgLTM1LjU4MzY5LC0zOC4xNzEyNCAtNTUuMzE2NCwtNjcu%0D%0AOTMxNjUgLTIwLjA1NjMsLTI5Ljc2MSAtMzUuMjU5OTcsLTY2Ljk2MTE1IC01NC4zNDU3LC0xMDUu%0D%0ANzc5MjkgQyAzOTYuMzQ1OTksNDI1LjY4ODgxIDM3MS40Mzc3LDM2Ni4xNjY5OCAzNDIuOTcwNywz%0D%0AMDYuNjQ2NDggMzEwLjI5ODQ1LDIyNy4wNjg0IDIzMS4wNDY5MywxMjUuODE2MDUgMjAxLjYwOTM4%0D%0ALDEyMS4yODcxMSBaIgogICAgICAgICAgIHRyYW5zZm9ybT0ibWF0cml4KC0xLjAyODc4NDIsMCww%0D%0ALC0xLjAyODc4NDIsMTg2OC4xNzc5LDY5NC4xMTA0NykiCiAgICAgICAgICAgaWQ9InBhdGg0MzUz%0D%0AIgogICAgICAgICAgIGlua3NjYXBlOmNvbm5lY3Rvci1jdXJ2YXR1cmU9IjAiIC8+CiAgICAgICAg%0D%0APHBhdGgKICAgICAgICAgICBzdHlsZT0iZmlsbDojZmZmZmZmO2ZpbGwtb3BhY2l0eToxO2ZpbGwt%0D%0AcnVsZTpub256ZXJvO3N0cm9rZTpub25lIgogICAgICAgICAgIGQ9Im0gOTcuMzI5MTYzLDUyNi40%0D%0ANDY2MSBjIDAsMy42NjA3NCAtMi45OTUyMDYsNi42NTU5MyAtNi42NTU5NTIsNi42NTU5MyAtMy42%0D%0ANjA3MzUsMCAtNi42NTU5MzUsLTIuOTk1MTkgLTYuNjU1OTM1LC02LjY1NTkzIDAsLTMuNjYwNzUg%0D%0AMi45OTUyLC02LjY1NTk0IDYuNjU1OTM1LC02LjY1NTk0IDMuNjYwNzQ2LDAgNi42NTU5NTIsMi45%0D%0AOTUxOSA2LjY1NTk1Miw2LjY1NTk0IHoiCiAgICAgICAgICAgaWQ9InBhdGg0MzU1IgogICAgICAg%0D%0AICAgIGlua3NjYXBlOmNvbm5lY3Rvci1jdXJ2YXR1cmU9IjAiIC8+CiAgICAgICAgPHBhdGgKICAg%0D%0AICAgICAgICBzdHlsZT0iZmlsbDojZmZmZmZmO2ZpbGwtb3BhY2l0eToxO2ZpbGwtcnVsZTpub256%0D%0AZXJvO3N0cm9rZTpub25lIgogICAgICAgICAgIGQ9Im0gMTM4LjEzNTEzLDU1Mi45ODQ2MiBjIDAs%0D%0AMy42NjA3MyAtMi42NjIzNCw2LjY1NTk0IC02LjY1NTkzLDYuNjU1OTQgLTMuNjYwNzQsMCAtNi42%0D%0ANTU5NSwtMi45OTUyMSAtNi42NTU5NSwtNi42NTU5NCAwLC0zLjY2MDkgMi45OTUyMSwtNi42NTYx%0D%0AMSA2LjY1NTk1LC02LjY1NjExIDMuOTkzNTksMCA2LjY1NTkzLDIuOTk1MjEgNi42NTU5Myw2LjY1%0D%0ANjExIHoiCiAgICAgICAgICAgaWQ9InBhdGg0MzU3IgogICAgICAgICAgIGlua3NjYXBlOmNvbm5l%0D%0AY3Rvci1jdXJ2YXR1cmU9IjAiIC8+CiAgICAgIDwvZz4KICAgICAgPGcKICAgICAgICAgdHJhbnNm%0D%0Ab3JtPSJtYXRyaXgoLTAuOTcyMDIxMTIsMCwwLC0wLjk3MjAyMTEyLDE4MTMuOTA4NCw1MjcuMDUy%0D%0AMjIpIgogICAgICAgICBpZD0iZzQzOTIiIC8+CiAgICA8L2c+CiAgPC9nPgo8L3N2Zz4K\" width=\"50%\"/>\n",
    "        </td>\n",
    "    </tr>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "<sup>The biopython logo [biopython license](https://github.com/biopython/biopython/blob/master/LICENSE.rst) © 2017 by Patrick Kunzmann. Source [biopython wiki](https://biopython.org/wiki/Logo)</sup>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[Return to top](#top)\n",
    "\n",
    "-------------\n",
    "\n",
    "<a id=\"install\"></a>\n",
    "\n",
    "<b><h2>2 Installing biopython</h2></b>  \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ensure the notebook is running in your virtual environment\n",
    "We want the following command to display something similar to\n",
    "\n",
    "`<PATH/TO>/anaconda3/envs/sprint/bin/python`\n",
    "\n",
    "The essential part of this the section `/sprint/bin/python` where **sprint** is the name of your new environment. If it looks like you are not in your new environment, refer to **week_4/INTRODUCTION.md**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! which python"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Install biopython and the Bio package\n",
    "*Note: preceding commands with \"!\" allows jupyter to activate command line functions*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We will use conda to install bopython rather than pip\n",
    "! conda install -c conda-forge biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Where possible it is generally recommended to install packages into conda environments using conda. Each [package](https://anaconda.org/anaconda/repo) is provided with installation commands "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[Return to top](#top)\n",
    "\n",
    "-------------\n",
    "\n",
    "\n",
    "<a id=\"overview\"></a>\n",
    "\n",
    "<b><h2>3 So what exactly is biopython</h2></b>  \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract from the [Biopython Tutorial and Cookbook](http://biopython.org/DIST/docs/tutorial/Tutorial.html)\n",
    "\n",
    "> The Biopython Project is an international association of developers of freely available Python (https://www.python.org) tools for computational molecular biology. Python is an object oriented, interpreted, flexible language that is becoming increasingly popular for scientific computing. Python is easy to learn, has a very clear syntax and can easily be extended with modules written in C, C++ or FORTRAN.\n",
    "\n",
    "> The Biopython web site (http://www.biopython.org) provides an online resource for modules, scripts, and web links for developers of Python-based software for bioinformatics use and research. Basically, the goal of Biopython is to make it as easy as possible to use Python for bioinformatics by creating high-quality, reusable modules and classes. Biopython features include parsers for various Bioinformatics file formats (BLAST, Clustalw, FASTA, Genbank,...), access to online services (NCBI, Expasy,...), interfaces to common and not-so-common programs (Clustalw, DSSP, MSMS...), a standard sequence class, various clustering modules, a KD tree data structure etc. and even documentation.\n",
    "\n",
    "> Basically, we just like to program in Python and want to make it as easy as possible to use Python for bioinformatics by creating high-quality, reusable modules and scripts.\n",
    "\n",
    "**Contributors**\n",
    "- Jeff Chang\n",
    "- Brad Chapman\n",
    "- Iddo Friedberg\n",
    "- Thomas Hamelryck\n",
    "- Michiel de Hoon\n",
    "- Peter Cock\n",
    "- Tiago Antao\n",
    "- Eric Talevich\n",
    "- Bartek Wilczyński\n",
    "\n",
    "Last Update – 16 July 2019 (Biopython 1.74)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[Return to top](#top)\n",
    "\n",
    "-------------\n",
    "\n",
    "<a id=\"bio_entrez\"></a>\n",
    "\n",
    "<b><h2>4 The Bio.Entrez Module</h2></b>  \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract from the [Biopython Tutorial and Cookbook](http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc109)\n",
    "\n",
    "> Entrez (https://www.ncbi.nlm.nih.gov/Web/Search/entrezfs.html) is a data retrieval system that provides users access to NCBI’s databases such as PubMed, GenBank, GEO, and many others. You can access Entrez from a web browser to manually enter queries, or you can use Biopython’s Bio.Entrez module for programmatic access to Entrez. The latter allows you for example to search PubMed or download GenBank records from within a Python script.\n",
    "\n",
    "> The Bio.Entrez module makes use of the Entrez Programming Utilities (also known as EUtils), consisting of eight tools that are described in detail on NCBI’s page at https://www.ncbi.nlm.nih.gov/books/NBK25501/. Each of these tools corresponds to one Python function in the Bio.Entrez module, as described in the sections below. This module makes sure that the correct URL is used for the queries, and that not more than one request is made every three seconds, as required by NCBI.\n",
    "\n",
    "> The output returned by the Entrez Programming Utilities is typically in XML format. To parse such output, you have several options:\n",
    "> 1. Use Bio.Entrez’s parser to parse the XML output into a Python object;\n",
    "> 2. Use the DOM (Document Object Model) parser in Python’s standard library;\n",
    "> 3. Use the SAX (Simple API for XML) parser in Python’s standard library;\n",
    "> 4. Read the XML output as raw text, and parse it by string searching and manipulation.\n",
    "\n",
    "\n",
    "### So what does this mean?\n",
    "In simple terms the Bio.Entrez module can be used to request data from the NCBI's Entrez APIs. The data are returned in a highly structured computer-readable format *e.g* XML ***(We will cover the XML format in greater detail in week 6 of this course)***, or human readable formats *e.g.* plain text\\*. A common usage of Bio.Entrez is downloading sequences in the FASTA or GenBank/GenPept plain text formats.\n",
    "\n",
    "\\* *NCBI changed the default return in Feb 2012, so many databases which previously returned text output now only give XML*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"seq\"></a>\n",
    "\n",
    "<b><h2>5 SeqIO, SeqRecord and Seq</h2></b>  \n",
    "</div>\n",
    "\n",
    "Biopython's standard Sequence [Input/Output interface](http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc47) [SeqIO](https://biopython.org/wiki/SeqIO) aims to provide a simple uniform interface to input and output assorted sequence file formats.\n",
    "\n",
    "In this tutorial we will use SeqIO to parse the data returned by Bio.Entrez into a structured Python object enabling us to access the components of the data. We will return data in the [Genbank flat file format](https://www.ncbi.nlm.nih.gov/Sitemap/samplerecord.html) ([SeqIO](https://biopython.org/wiki/SeqIO) can parse records  in several different formats - see the provided link).\n",
    "\n",
    "The data are formatted by SeqIO into a [SeqRecord](https://biopython.org/wiki/SeqRecord) object. The sequence itself is held in a [Seq](https://biopython.org/wiki/Seq) object. The Seq object has two attributes (record.seq):\n",
    "1. The sequence - my_seq.seq\n",
    "2. The Alphabet - my_seq.alphabet\n",
    "\n",
    "The sequence attribute is pretty self explanatory, but what about the alphabet. The alphabet comprises the valid International Union of Pure and Applied Chemistry [IUPAC](https://iupac.org/) characters which make up the sequence, *e.g.* the DNAAlphabet (ATGC) RNAAlphabet(aucc) and ProteinAlphabet. Biopython also provides alphabets for [extended alphabets](https://biopython.org/DIST/docs/api/Bio.Alphabet.IUPAC-module.html), *e.g.* enabling characters such as the IUPAC symbol N (any DNA base) or X (any amino acid - ***which is sometimes used illegally to specify protein termination, which is correctly specified by the IUPAC character \\**** ).\n",
    "\n",
    "In addition, the [Seq](https://biopython.org/wiki/Seq) object contains useful sequence associated methods including:\n",
    "\n",
    "1. Complement and reverse-complement\n",
    "```python\n",
    ">>> from Bio.Seq import Seq\n",
    ">>> from Bio.Alphabet import generic_dna\n",
    ">>> my_dna = Seq(\"AGTACACTGGT\", generic_dna)\n",
    ">>> my_dna\n",
    "Seq('AGTACACTGGT', DNAAlphabet())\n",
    ">>> my_dna.complement()\n",
    "Seq('TCATGTGACCA', DNAAlphabet())\n",
    ">>> my_dna.reverse_complement()\n",
    "Seq('ACCAGTGTACT', DNAAlphabet())\n",
    "```\n",
    "2. Transcription and reverse-transcription\n",
    "```python\n",
    ">>> my_rna = my_dna.transcribe()\n",
    ">>> my_rna\n",
    "Seq('AGUACACUGGU', RNAAlphabet())\n",
    ">>> my_rna.back_transcribe()\n",
    "Seq('AGTACACTGGT', DNAAlphabet())\n",
    "```\n",
    "3. Translation\n",
    "```python\n",
    ">>> from Bio.Seq import Seq\n",
    ">>> from Bio.Alphabet import generic_dna\n",
    ">>> coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\", generic_dna)\n",
    ">>> coding_dna.translate()\n",
    "Seq('MAIVMGR*KGAR*', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
    ">>> coding_dna.translate(to_stop=True)\n",
    "Seq('MAIVMGR', ExtendedIUPACProtein())\n",
    "```\n",
    "\n",
    "\n",
    "The remaining attributes of the [SeqRecord](https://biopython.org/wiki/SeqRecord) object provide additional information about the a parsed record, *e.g.*\n",
    "\n",
    "```python\n",
    ">>> record.id\n",
    "\"NM_000059.3\"\n",
    ">>> record.description\n",
    "\"Homo sapiens BRCA2 DNA repair associated (BRCA2), mRNA\"\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"be_example\"></a>\n",
    "\n",
    "<b><h2>6 Bio.Entrez, SeqIO and Seq example</h2></b>  \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this introductory tutorial we will use the Bio.Entrez module to return the human [BRCA2 BRCA2 DNA repair associated](https://www.ncbi.nlm.nih.gov/gene/675) reference sequence record [NM_000059.3 Homo sapiens BRCA2 DNA repair associated (BRCA2), mRNA](https://www.ncbi.nlm.nih.gov/nuccore/NM_000059.3) in the Genbank flat file format. We will then use SeqIO to parse the record and create a SeqRecord object with the sequence contained in a Seq object. We will then look at some of the SeqIO attributes and Seq methods.\n",
    "\n",
    "\n",
    "\n",
    "#### 1. Import and configure the Bio.Entrez module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# import Bio modules\n",
    "from Bio import Entrez, SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify yourself so the NCBI can make contact if there are any issues (see Biopython Eutils guidance - below)\n",
    "Entrez.email = '<YOUR EMAIL ADDRESS>' # Note: change this to your own email address!!!!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. Fetch the RefSeq record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch the sequence record from RefSeq. Note: db is the database and gb is a Genbank record\n",
    "handle = Entrez.efetch(db=\"nucleotide\", id='NM_000059.3', rettype=\"gb\", retmode=\"text\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3. Parse and close the handle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse the handle into a SeqRecord object called record\n",
    "record = SeqIO.read(handle, \"gb\")\n",
    "# Close the handle\n",
    "handle.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4. Explore the SeqRecord object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# List the SeqRecord attributes using Python's dir() function\n",
    "dir(record)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "solution": "hidden",
    "solution_first": true
   },
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "### Exercise 1\n",
    "\n",
    "We are not particularly interested in the attributes prefixed `_` at this stage. \n",
    "\n",
    "Write simple code including a for loop that only prints the attributed not prefixed `_`\n",
    "\n",
    "*Hint: you will need to use Python's Regular Expression moduel [re](https://www.w3schools.com/python/python_regex.asp)*\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "solution": "hidden"
   },
   "source": [
    "**Solution**\n",
    "\n",
    "```python\n",
    "import re\n",
    "attribs = dir(record)\n",
    "for at in attribs:\n",
    "    if re.match('^_', at):\n",
    "        continue\n",
    "    else:\n",
    "        print(at)   \n",
    " ```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Your code here\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5. Begin extracting data from the SeqRecord"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "solution": "hidden",
    "solution_first": true
   },
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "### Exercise 2\n",
    "\n",
    "The structure of the SeqRecord, which we have named `record`, is extremely useful, but it contains a wealth of information that we do not need to store and work with. Also, some of the sub-objects such as the Seq object (`record.seq`) contain methods that are useful but become redundant if we want to store or print simple [serializable](https://en.wikipedia.org/wiki/Serialization) data.\n",
    "\n",
    "Your task is to begin adding data from the SeqRecord to a [Python dictionary](https://www.w3schools.com/python/python_dictionaries.asp).\n",
    "\n",
    "\n",
    "First create a dictionary with the following structure and populate it with data from the SeqRecord\n",
    "\n",
    "```python\n",
    "d1 = {'id': \"<The accession number of the record, i.e. NM_123456.7>\",\n",
    "      'description': \"<My Favourite Gene (MFG), mRNA>\"}\n",
    "```\n",
    "\n",
    "***Remember to print test the dictionary by printing the values***\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "solution": "hidden"
   },
   "source": [
    "**Solution**\n",
    "```python\n",
    "d1 = {'id': record.id,\n",
    "      'description': record.description}\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Your code here\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "solution": "hidden"
   },
   "outputs": [],
   "source": [
    "# Test your dictionary\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 6. Extracting and creating data using the Seq object"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "solution": "hidden",
    "solution_first": true
   },
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "### Exercise 3\n",
    "It is not necessarily obvious, but transcript records use the IUPAC DNA alphabet rather than the RNA alphabet. The simple reason for this is that transcript records are derived from sequencing reverse-transcribed cDNA.\n",
    "\n",
    "Using the the [Seq](https://biopython.org/wiki/Seq) objects in-built methods add the following keys to your dictionary and populate the values\n",
    "\n",
    "```python\n",
    "d1['cdna'] = \"The cDNA sequence of the record (note: the string only, not the Seq object)\"\n",
    "d1['rna'] = \"The RNA sequence of the record (note: we want lower-case RNA bases)\"\n",
    "```\n",
    "\n",
    "*Note: you can list the Seq object attributes using*\n",
    "```python\n",
    "dir(record.seq)\n",
    "```\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "solution": "hidden"
   },
   "source": [
    "**Solution**\n",
    "```python\n",
    "# assign record.seq to an object\n",
    "my_cdna_obj = record.seq\n",
    "# extract the cDNA sequence \n",
    "my_cdna = str(record.seq)\n",
    "# re-create the Seq object in RNA format\n",
    "my_rna_obj = record.seq.transcribe()\n",
    "# extract the RNA sequence, and lower-case \n",
    "my_rna = str(my_rna_obj.lower())\n",
    "# add the data to the dictionary\n",
    "d1['cdna'] = my_cdna\n",
    "d1['rna'] = my_rna\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Your code here\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "solution": "shown"
   },
   "source": [
    "#### 7. Features - Gene and translation information\n",
    "The SeqRecord features are annotations relating to the reference sequence record. They are not particularly intuative to use so we will concentrate on extracting useful information from the SeqRecord.features which comprises a list of [SeqFeature](http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc38) objects\n",
    "\n",
    "To our dictionary we want to add the following items\n",
    "\n",
    "```python\n",
    "d1['gene']['gene_symbol'] = 'BRCA2'\n",
    "d1['gene']['hgnc_id'] = 'HGNC:1101'\n",
    "d1['gene']['entrez_id'] = '675'\n",
    "```\n",
    "***Note: We have created a nested dictionary key-worded 'gene'. This allows us to separate out the gene-level data from a data-structure that is primarily transcript focused. It also provides us with a simple method to print the gene-level information we have gathered about the transcript. Similarly, we will separate out the translation into a nested dictionary below***\n",
    "\n",
    "We will also go on to add protein information to the dictionary\n",
    "\n",
    "```python\n",
    "d1['type'] = cds # Note: non-coding RNAs have the type ncrna (ncRNA)\n",
    "d1['translation']['cds_start'] = 227 # Note: ncRNA don't have protein information so set to None\n",
    "d1['translation']['cds_end'] = 10484\n",
    "d1['translation']['id'] = 'NP_000050.2'\n",
    "d1['translation']['description'] = 'breast cancer type 2 susceptibility protein'\n",
    "d1['translation']['prot'] = \"<the protein sequence>\"\n",
    "```\n",
    "\n",
    "Truthfully, the SeqFeature object is not particularly intuitive, and the documentation is not necessarily as helpful as it could be. However, luckily all the information we want at this stage can be accessed through a single SeqFeture object\n",
    "\n",
    "Run the following code to identify the correct element and print the relevant information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SeqRecord.features is a list of SeqFeature objects\n",
    "element = 0\n",
    "for feature in record.features:\n",
    "    if feature.type == 'CDS': # Note, a non-coding transcript would have the type ncRNA\n",
    "        report = 'Printing record.features[%s]' % str(element)\n",
    "        print(report)\n",
    "        print(record.features[element])\n",
    "        break\n",
    "    else:\n",
    "        element = element + 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "solution": "hidden",
    "solution_first": true
   },
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "### Exercise 4\n",
    "\n",
    "Populate the gene data in dictionary d1\n",
    "\n",
    "```python\n",
    "d1['gene']['gene_symbol'] = 'BRCA2'\n",
    "d1['gene']['hgnc_id'] = 'HGNC:1101'\n",
    "d1['gene']['entrez_id'] = '675'\n",
    "```\n",
    "\n",
    "Useful Python functions and information:\n",
    "- [re.search](https://www.w3schools.com/python/python_regex.asp#search)\n",
    "- [Regular Expression cheat-sheet](https://www.debuggex.com/cheatsheet/regex/python)\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "solution": "hidden"
   },
   "source": [
    "**Solution**\n",
    "\n",
    "```python\n",
    "\"\"\"\n",
    "import modules\n",
    "\n",
    "json is a standard serialized data format (as is XML)\n",
    "We will cover the use of json in this module, but for now I want to use Python's json module to provide a \"pretty\" way to print a Python dictionary\n",
    "\"\"\"\n",
    "import json\n",
    "import re\n",
    "\n",
    "# create the gene sub-dictionary within d1\n",
    "d1['gene'] = {}\n",
    "\n",
    "# Add the gene symbol. I don't know why the biopython folks chose to use a list, \"there can be only one\"!\n",
    "gene_symbol = record.features[5].qualifiers['gene'][0]\n",
    "d1['gene']['gene_symbol'] = gene_symbol\n",
    "\n",
    "\"\"\"\n",
    "Use a regular expression search to identify the gene_symbol\n",
    "re.search(r'\\w+', STRING) searches for any run of alphanumeric characters in a string\n",
    "\n",
    "I could have also looped through the record.features[5].qualifiers['db_xref'] \n",
    "list until I found re.match(HGNC, <element>)\n",
    "\n",
    "Again, there \"can be only one\" HGNC id, so we can assume re.search will only return one result. \n",
    "Therefore, I have split the 0th element of the re.search at the ':' character recovering the last element\n",
    "i.e. [0].split(':')[-1]\n",
    "\n",
    "I then put-back the preceding \"HGNC:\" because I have only recovered the numbers from HGNC:HGNC:1101\n",
    "\"\"\"\n",
    "hgnc_id = (re.search(r'HGNC:HGNC:\\d+', str(record.features[5].qualifiers['db_xref'])))[0].split(':')[-1]\n",
    "hgnc_id = 'HGNC:%s' %hgnc_id\n",
    "d1['gene']['hgnc_id'] = hgnc_id\n",
    "\n",
    "\"\"\"\n",
    "I apply the same logic to the entrez ID, but this time I only want to recover the numbers. \n",
    "\n",
    "Again, searching through a loop would be valid\n",
    "\"\"\"\n",
    "entrez_id = (re.search(r'GeneID:\\d+', str(record.features[5].qualifiers['db_xref'])))[0].split(':')[-1]\n",
    "d1['gene']['entrez_id'] = entrez_id\n",
    "\n",
    "# print the resulting dictionary using json.dumps\n",
    "print(json.dumps(d1['gene'], sort_keys=True, indent=4, separators=(',', ': ')))\n",
    "```\n",
    "\n",
    "**The code should return**\n",
    "```python\n",
    "{\n",
    "    \"entrez_id\": \"675\",\n",
    "    \"gene_symbol\": \"BRCA2\",\n",
    "    \"hgnc_id\": \"HGNC:1101\"\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Your code here\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "solution": "hidden",
    "solution_first": true
   },
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "### Exercise 5\n",
    "\n",
    "Now Populate the translation data in dictionary d1\n",
    "\n",
    "```python\n",
    "d1['type'] = cds # Note: non-coding RNAs have the type ncrna (ncRNA)\n",
    "d1['translation']['cds_start'] = 227 # Note: ncRNA don't have protein information so set to None\n",
    "d1['translation']['cds_end'] = 10484\n",
    "d1['translation']['id'] = 'NP_000050.2'\n",
    "d1['translation']['description'] = 'breast cancer type 2 susceptibility protein'\n",
    "d1['translation']['prot'] = \"<the protein sequence>\"\n",
    "```\n",
    "\n",
    "Useful Python functions and information:\n",
    "- [re.search](https://www.w3schools.com/python/python_regex.asp#search)\n",
    "- [Regular Expression cheat-sheet](https://www.debuggex.com/cheatsheet/regex/python)\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "solution": "hidden"
   },
   "source": [
    "**Solution**\n",
    "\n",
    "The protein dictionary is much easier to populate since all but the CDS fields have single values. Again, I'm unsure why the biopython teams chose to put single strings into lists!\n",
    "\n",
    "```python\n",
    "# The record type in this instance is a CDS - but we want to store this a lower.case\n",
    "d1['type'] = record.features[5].type.lower()\n",
    "\n",
    "# Create the translation sub-dictionary\n",
    "d1['translation'] = {}\n",
    "\n",
    "\"\"\"\n",
    "The cds start and end positions are stored in an attribute labeled location\n",
    "The structure is a text string \"[227:10484](+)\"\n",
    "\n",
    "I have chosen to perform a regular expression search for \"<string-of-numbers>:<string-of-numbers>\" or re.search(r\"\\d+:\\d+\", <STRING>)\n",
    "\n",
    "The resulting string can be split straight into the dictionary at the \":\" character\n",
    "\"\"\"\n",
    "cds_info = re.search(r\"\\d+:\\d+\", str(record.features[5].location))[0]\n",
    "d1['translation']['cds_start'], d1['translation']['cds_end'] = cds_info.split(\":\")\n",
    "\n",
    "# All other items require identification and converting from a single-element list into a string\n",
    "d1['translation']['id'] = record.features[5].qualifiers['protein_id'][0]\n",
    "d1['translation']['description'] = record.features[5].qualifiers['product'][0]\n",
    "d1['translation']['prot'] = record.features[5].qualifiers['translation'][0]\n",
    "\n",
    "# Finally print the pretty dictionary!\n",
    "print(json.dumps(d1['translation'], sort_keys=True, indent=4, separators=(',', ': ')))\n",
    "```\n",
    "\n",
    "**The output should look like this**\n",
    "\n",
    "```python\n",
    "{\n",
    "    \"cds_end\": \"10484\",\n",
    "    \"cds_start\": \"227\",\n",
    "    \"description\": \"breast cancer type 2 susceptibility protein\",\n",
    "    \"id\": \"NP_000050.2\",\n",
    "    \"prot\": \"MPI ... KYI\" # I have shortened the sequence for neatness\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "solution": "hidden"
   },
   "outputs": [],
   "source": [
    "# Your code here\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[Return to top](#top)\n",
    "\n",
    "-------------\n",
    "\n",
    "\n",
    "<a id=\"sanity\"></a>\n",
    "\n",
    "<b><h2>7 Sanity checking with biopython</h2></b>  \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have used Bio.Entrez to recover a transcript record from RefSeq, parsed it into a SeqIO object and recovered useful information from this highly-complex object-of-objects into a simple Python dictionary.\n",
    "\n",
    "However, the protein information has been derived from the transcript reference sequence, so can we:\n",
    "\n",
    "1. Replicate the translation\n",
    "2. Cross check the translation with the identified protein reference sequence record"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "solution": "hidden",
    "solution_first": true
   },
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "### Exercise 1\n",
    "\n",
    "1. Create a new [Seq](https://biopython.org/wiki/Seq) object from the RNA sequence stored in your dictionary and use the Seq object's built in functions to translate the RNA sequence.\n",
    "\n",
    "2. [Translate](https://biopython.org/DIST/docs/api/Bio.Seq.Seq-class.html#translate) the sequence specifying `cds=True` and `to_stop=True`\n",
    "\n",
    "3. Compare the translation to `d1['translation']['prot']`. They ought to be identical\n",
    "\n",
    "*Note: we know the sequence is an RNA sequence, so make sure you tell Bio.Seq*\n",
    "\n",
    "*Hint: You will need to consider the \"actual\" sequence that will be translated (CDS). Will you want to store it?*\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "solution": "hidden"
   },
   "source": [
    "**Solution**\n",
    "\n",
    "```python\n",
    "# import tho Bio.Seq Seq object\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# Import a relevant alphabet from Bio.alphabet. I have chosen the IUPAC alphabet object\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n",
    "# Assign the RNA sequence from the dictionary to a variable\n",
    "my_rna = d1['rna']\n",
    "\n",
    "# Extract the CDS. Note: UTRs do not get translated\n",
    "my_cds = my_rna[int(d1['translation']['cds_start']):int(d1['translation']['cds_end'])]\n",
    "\n",
    "# Create a Seq object for the CDS\n",
    "my_cds_obj = Seq(my_cds, IUPAC.unambiguous_rna)\n",
    "\n",
    "# I'm using Seq.translate in CDS mode. This allows me to determine whether there is a valid translation initiation \n",
    "# codon at positions 1 to 3 of the CDS and a valid stop codon at the positions -3 to -1 of the CDS\n",
    "# Note: Not all human CDS contain AUG at positions 1 to 3. More information about translations tables is available in\n",
    "# the Translate link above\n",
    "my_trans_obj = my_cds_obj.translate(table=1, cds=True, to_stop=True)\n",
    "\n",
    "# Capture the translation string and compare it to the recovered string stored in the dictionary\n",
    "trans = str(my_trans_obj)\n",
    "if trans == d1['translation']['prot']:\n",
    "    print(\"True\")\n",
    "else:\n",
    "    print(\"False\")\n",
    "\n",
    "# Since I'm happy with the translation, I know the CDS is correct, so I might as well store it\n",
    "d1['translation']['prot'] = my_cds\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Your code here\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "### Exercise 2\n",
    "\n",
    "Protein records can also be recovered from RefSeq.\n",
    "\n",
    "Use a combination of the steps above to recover the correct protein record from RefSeq and ensure the translated sequence is identical to the translation provided in the transcript record\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Your code here\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[Return to top](#top)\n",
    "\n",
    "-------------\n",
    "\n",
    "\n",
    "<a id=\"summary\"></a>\n",
    "\n",
    "<b><h2>8 Summary and Assignment</h2></b>  \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this tutorial we have introduced the concept of using the biopython Bio.Entrez module to query the NCBI RefSeq database via their Entrez API. We have recovered RefSeq reference sequence records in the Genbank format, and parsed them into a Bio.SeqRecord object using Bio.SeqIO. We have explored the contents of the SeqRecord object including the Bio.Seq object and the Bio.Features object. \n",
    "\n",
    "As previously stated, we can only scratch the surface of biopython's functionality in this module. In truth, we can only scratch the surface of the powerful searches that can be performed using Bio.Entrez"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "cell_style": "split"
   },
   "source": [
    "### Over to you\n",
    "\n",
    "#### Aim of this exercise\n",
    "The aim of this exercise is to get you into the mindset of working together as a team. We will concentrate on aspects of working in an Agile fashion.\n",
    "\n",
    "#### Structure your team\n",
    "You will need to allocate the following team roles:\n",
    "\n",
    "1. **Project lead**\n",
    "    - Initiate the project on Git Issues\n",
    "    - Lead the group discussion in Git Issues and Slack\n",
    "    - Provide final feedback on the group's activities and close the issue\n",
    "    \n",
    "\n",
    "2. **Team members**\n",
    "    - Coders who will be responsible for writing the Python functions\n",
    "    - Testers who will be responsible for testing the code and providing feedback to the coders\n",
    "\n",
    "***We recommend ensuring that you most experienced coders work with your least experienced coders. Don't forget, this is a team assignment, if you can't figure out how to do something, ask your team on Slack!***\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "cell_style": "split"
   },
   "source": [
    "<img src=\"https://github.com/i3hsInnovation/resources/blob/master/images/Discussionsummative.png?raw=true\" width=\"25%\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "cell_style": "split"
   },
   "source": [
    "### Work-flow\n",
    "\n",
    "1. Group leader creates an issue on Git Issues\n",
    "2. The coders will work together to write the module\n",
    "3. The testers will review the final code and test the code. Feedback will be given to the coders within the Git issue\n",
    "4. Once the coding is completed and tested, the project lead will summarise the key work-flow points and close the issue\n",
    "\n",
    "**Details about how the assignment will be marked can be found [here](LINK)**\n",
    "\n",
    "#### Considerations\n",
    "1. How will the coders use Git to share the coding tasks?\n",
    "2. How will the testers be asked to review the code via Git issues?\n",
    "3. Before you start coding, as a team research how to use branching and pull-requests, for example, my working code is generally pushed to a branch called `develop` before being merged into `master`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "### Group Assignment\n",
    "\n",
    "#### Setup and general workflow\n",
    "First the project leader will create a branch for this SPRINT and then create a directory in the level above this directory called **SPRINT**, *i.e.* `Introduction-to-Programming/SPRINT`. The leader will add the directory so that it is tracked in Git and push the directory to a remote branch in GitHub. The coders can then checkout the remote branch and begin coding.\n",
    "\n",
    "<br>\n",
    "\n",
    "Create a local branch of the repo\n",
    "```bash\n",
    "$ git branch <branch_name>\n",
    "$ git checkout <branch_name>\n",
    "```\n",
    "Complete the coding then add **ONLY** the SPRINT directory\n",
    "```bash\n",
    "$ git add SPRINT\n",
    "```\n",
    "Commit your changed and push\n",
    "```bash\n",
    "$ git commit -m \"A useful message\"\n",
    "$ git push -u origin <branch_name>\n",
    "```\n",
    "\n",
    "Checkout a remote branch\n",
    "```bash\n",
    "$ git checkout --track origin/<branch_name>\n",
    "```\n",
    "\n",
    "#### Tasks\n",
    "Below are two Python functions:\n",
    "1. get_entrez_gene_id - Returns the Entrez Gene ID when provided with a valid HGNC gene symbol\n",
    "2. get_transcript_list - Returns a list of transcripts (coding and non-coding excluding model transcripts) for the stated gene\n",
    "\n",
    "Your task is to write a Python module in the directory **SPRINT** directory containing a function that creates and returns a list of dictionaries (as we created above) for each and every [non-model](https://www.ncbi.nlm.nih.gov/books/NBK50679/#RefSeqFAQ.what_is_the_difference_between) transcript for the ***BRCA1*** gene.\n",
    "\n",
    "Considerations:\n",
    "- The structure of the list should be, for example:\n",
    "\n",
    "    ```python\n",
    "    [{........},\n",
    "     {........},\n",
    "     {........}]\n",
    "    ```\n",
    "- The overall structure of the flat text file must be dictionary-like. We recommend [json.dumps](https://www.dev2qa.com/python-json-dump-to-or-load-from-file-example/). *Note: Json will be covered in more detail in the next tutorial.* \n",
    "\n",
    "- It is entirely up to you how the module is activated. Usual options are to\n",
    "\n",
    "a. You may choose to trigger a script directly\n",
    "```bash\n",
    "python a_script.py\n",
    "```\n",
    "b. You may wish to create a module that can be imported into the Python terminal\n",
    "\n",
    "c. You may wish to create a module which is  usable in both of these ways\n",
    "\n",
    "### Team responsibilities\n",
    "It is the responsibility of the entire team to plan the project and decide how the module will work. Use Slack for your discussions and use Git issues to document your decisions and how to assign the individual tasks to each team member. Use a single Git Issue with multiple comments rather than several issues!\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "solution": "hidden"
   },
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "import re\n",
    "Entrez.email = \"<YOUR EMAIL ADDRESS>\"  # Always tell NCBI who you are\n",
    "\n",
    "\"\"\"\n",
    "Function which returns the Entrez gene id when provided with a valid HGNC gene symbol\n",
    "\"\"\"\n",
    "def get_entrez_gene_id(gene_symbol):\n",
    "    # Create a search term. We are searching for a human[orgn] gene[gene] \n",
    "    term = '%s[gene] \"Homo sapiens\"[orgn]' % gene_symbol\n",
    "    \n",
    "    # Request data from Entrez using Bio.Entrez\n",
    "    handle = Entrez.esearch(db='gene', \n",
    "                            term=term, \n",
    "                            retmode='xml'\n",
    "                           )\n",
    "    \n",
    "    # The data are returned in the xml format (which will be covered in a later tutorial),\n",
    "    # and converted into a Python dictionary using the Entrez.read module\n",
    "    record = Entrez.read(handle)\n",
    "    \n",
    "    # Close the handle\n",
    "    handle.close()\n",
    "    \n",
    "    # Return the gene id which is stores as a list element keyworded 'IdList'\n",
    "    # There should only ever be 1 gene ID for each gene!\n",
    "    return (record['IdList'][0])\n",
    "\n",
    "# Test the function - you can run the cell to activate this code\n",
    "entrez_id = (get_entrez_gene_id('BRCA1'))\n",
    "entrez_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Function which returns the valid transcript reference sequences when provided with an Entrez gene ID\n",
    "\"\"\"\n",
    "def get_transcript_list(entrez_id, gene_table=False):\n",
    "    \n",
    "    # Request data from Entrez using Bio.Entrez\n",
    "    handle = Entrez.efetch(db=\"gene\",\n",
    "                       id=entrez_id,\n",
    "                       rettype=\"gene_table\", # We are returning a gene_table\n",
    "                       retmode=\"text\"\n",
    "                      )\n",
    "    \n",
    "    # We have returned text so that it can be easily manipulated. We read the text as if it were a file\n",
    "    record = handle.read()\n",
    "    \n",
    "    # Close the handle\n",
    "    handle.close()\n",
    "    \n",
    "    # Optional request for the gene_table rather than the transcript list\n",
    "    if gene_table is True:\n",
    "        return record\n",
    "    \n",
    "    # Search the text for all coding transcripts using re.findall - returns a list\n",
    "    coding = re.findall(r\"NM_\\d+.\\d+\", record)\n",
    "    # Search the text for all non-coding transcripts using re.findall - returns a list\n",
    "    noncoding = re.findall(r\"NR_\\d+.\\d+\", record)\n",
    "    \n",
    "    # Combine the list\n",
    "    tx_list = coding + noncoding\n",
    "    \n",
    "    # Remove duplicate items - see * below\n",
    "    tx_list = list(dict.fromkeys(tx_list))\n",
    "    \n",
    "    # Return the transctipt list\n",
    "    return tx_list\n",
    "\n",
    "# Test the function\n",
    "print(get_transcript_list(str(entrez_id)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\**How to [remove duplicate items](https://www.w3schools.com/python/python_howto_remove_duplicates.asp) from a list*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A point of interest\n",
    "You are probably not aware, but by printing your final dictionary to a file you have effectively created a very small and very simple database. Because you have used JSON (which we will discuss in the next notebook, you have also created a database which will be compatible with modern [NoSQL](https://www.mongodb.com/nosql-explained) databases such as [MongoDB](https://www.mongodb.com/what-is-mongodb) which are becoming increasingly popular for storing bioinformatics data. Also, because your data have been compartmentalised, *i.e.* gene level information, the data could be separated into an additional data-structure and imported into your transcript record by referring to it's unique object identifier, or the data could be displayed separately to your transcript record. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Notebook details\n",
    "<br>\n",
    "<i>Notebook created by <strong>Dr Peter Causey-Freeman</strong> with <strong>Frances Hooley</strong> \n",
    "    \n",
    "\n",
    "Publish date: October 2020<br>\n",
    "Review date: October 2021</i>\n",
    "\n",
    "Please give your feedback using the button below:\n",
    "\n",
    "<a class=\"typeform-share button\" href=\"https://form.typeform.com/to/YMpwLTNy\" data-mode=\"popup\" style=\"display:inline-block;text-decoration:none;background-color:#3A7685;color:white;cursor:pointer;font-family:Helvetica,Arial,sans-serif;font-size:18px;line-height:45px;text-align:center;margin:0;height:45px;padding:0px 30px;border-radius:22px;max-width:100%;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;font-weight:bold;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;\" target=\"_blank\">Rate this notebook </a> <script> (function() { var qs,js,q,s,d=document, gi=d.getElementById, ce=d.createElement, gt=d.getElementsByTagName, id=\"typef_orm_share\", b=\"https://embed.typeform.com/\"; if(!gi.call(d,id)){ js=ce.call(d,\"script\"); js.id=id; js.src=b+\"embed.js\"; q=gt.call(d,\"script\")[0]; q.parentNode.insertBefore(js,q) } })() </script>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
